{% extends "_layouts/default.html" %}

{% block custom_head %}

<!-- jQuery UI CSS -->
<link rel="stylesheet" type="text/css"
      href="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/themes/smoothness/jquery-ui.css"/>

<!-- IGV CSS -->
<link rel="stylesheet" type="text/css" href="//igv.org/web/release/1.0.9/igv-1.0.9.css">

<!-- jQuery JS -->
<script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/jquery-ui.min.js"></script>
<script>$.widget.bridge('uitooltip', $.ui.tooltip);</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.6/js/bootstrap.min.js" integrity="sha256-KXn5puMvxCw+dAYznun+drMdG1IFl3agK0p/pqT9KAo=" crossorigin="anonymous"></script>

<script type="text/javascript" src="//igv.org/web/release/1.0.9/igv-1.0.9.min.js"></script>

<style>

.label {
  width: 80px;
  height: 18px;
  line-height: 12px;
  margin-bottom: 4px;
  display: inline-block;
}

.gt-3 {
    background-color: white;
    border: 1px dotted black;
    color: black;
}

.gt-0.PASS {
  background-color: rgba(194,194,214,1.0);
  border: 1px solid black;
  color: black;
}

.gt-2.PASS {
  background-color: rgba(0, 102, 255,1.0);
  border: 1px solid black;
  color: white;
}

.gt-0:not(.PASS) {
  background-color: rgba(194,194,214,0.25);
  border: 1px dotted black;
  color: black;
}

.gt-2:not(.PASS) {
  background-color: rgba(0, 102, 255,0.25);
  border: 1px dotted black;
  color: black;
}

.het {
  background-color: #ffff00;
  color: black;
}

.gt_set {
  border-right: 1px dotted #b3b3b3;
}

th {
  white-space: nowrap;
}

#variants {
  font-size: 12px;
}

</style>

{% endblock %}


{% block content %}
<div class="row">
      <div class="col-md-12">
    <div class="well">
    <div class="row">
      <div class="col-md-2">
      <h5>Tracks <a href="{{ url_for('primary.help_item', filename = 'Variant-Prediction') }}#standard-tracks"><span class="glyphicon glyphicon-question-sign"></span></a></h5>
      {% for i in ["Genes", "Transcripts"] %}
        <div class="checkbox">
          <label><input type="checkbox" checked class="track-select normal-track" value="{{ i }}" />{{ i }}</label>
        </div>
      {% endfor %}
      {% for i in ["phyloP", "phastCons", "Variants"] %}
        <div class="checkbox">
          <label><input type="checkbox" class="track-select normal-track" value="{{ i }}" />{{ i }}</label> 
        </div>
      {% endfor %}

        <h5>Variant Impact <a href="{{ url_for('primary.help_item', filename = 'Variant-Browser') }}#variant-effects"><span class="glyphicon glyphicon-question-sign"></span></a></h5>


        <div class="checkbox">
              <label><input type="checkbox" class="track-select variant-impact" value="ALL" />No Annotation <div class="legend-box" style="background-color: #fff;"></div></label>
        </div>

        <div class="checkbox">
              <label><input type="checkbox" class="track-select variant-impact" value="MODIFIER" />MODIFIER <div class="legend-box" style="background-color: #6f6f6f;"></div></label>
        </div>

        <div class="checkbox">
            <label><input type="checkbox" class="track-select variant-impact" value="LOW" />LOW <div class="legend-box" style="background-color: #66d866;"></div></label>
        </div>

        <div class="checkbox">
              <label><input type="checkbox" class="track-select variant-impact" value="MODERATE" checked />MODERATE <div class="legend-box" style="background-color: #ffd33f;"></div></label>
        </div>

        <div class="checkbox">
              <label><input type="checkbox" class="track-select variant-impact" value="HIGH" checked />HIGH <div class="legend-box" style="background-color: #ff3f3f;"></div></label>
        </div>


      </div>
      <div class="col-md-6">

      <h5>Gene Search</h5>
        <input type="text" class="form-control" id="gene-search" placeholder="Gene Search (e.g. trt-1)">
      <div id="loading-search-table" style="display:none; margin-top:20px;"><div class="spinner"><div class="bounce1"></div><div class="bounce2"></div><div class="bounce3"></div></div></div>
      
      <table id="search-table" class='table table-striped' style="display: none;">
      <thead>
      <tr>
        <th><em>C. elegans Gene</em></th>
        <th>Symbol/ID</th>
        <th>Species</th>
        <th>Source</th>
      </tr>
      </thead>
    <tbody id="orthologs"></tbody>
    </table>

      </div>

      <div class="col-md-4">
        <h5>Strains <a href="{{ url_for('primary.help_item', filename = 'Variant-Browser') }}#ind-strains"><span class="glyphicon glyphicon-question-sign"></span></a></h5>

        <div class="checkbox">
            <label><input type="checkbox" class="list-all-strains" value="list-all-strains" />List all strain genotypes </label>
        </div>

        <form class="form-inline pull-xs-right">

          <input class="form-control" id="filter" type="text" style="width:100%;  margin-bottom: 5px;" placeholder="e.g. JU360" autocomplete="off">

        </form>

        <table id='browser_strain_list' class="table table-striped table-hover table-condensed">
          
          <tbody class="searchable">

            {% for isotype in isotype_listing %}

            <tr>
              <td>
                {{ isotype }}
              </td>
              <td>
                  <label>
                  <input type="checkbox" class="track-select isotype-item sample-track" value="{{ isotype }}" />
                  VCF</label>
              </td>
              <td>
                  <label>
                  <input type="checkbox" class="track-select sample-track-alignment" value="{{ isotype }}_bam" />
                  BAM</label>
              </td>
            </tr>

            {% endfor %}

          </tbody>

        </table>
        </div>
      </div>

     
    </div>
</div>{# Close col #}
</div>{# Close row #}


<div id="browser"></div>

<div id="row" id="browser-legend">

    <div class="col-lg-12">
        <div class="legend-box" style="background-color: #c2c2d6;"></div> <strong>Reference</strong>&nbsp;&nbsp;
        <div class="legend-box" style="background-color: #0066ff;"></div> <strong>Alternate</strong>
          <label><div class="legend-box" style="background-color: #66d866;"> </div> LOW</label>&nbsp;&nbsp;
        <label><div class="legend-box" style="background-color: #ffd33f;"></div> MODERATE</label>
        <label><div class="legend-box" style="background-color: #ff3f3f;"></div> HIGH</label>
    </div>

</div>


<div class="row">
<div class="col-md-12 col-lg-12">
<h3>Variants</h3>


<button id='download' class='btn btn-primary btn-xs pull-right'>Download</button>
<p><small>A Maximum of 1000 variants from the region will be listed. Heterozygous calls are likely errors. When rows are yellow, it indicates the entire variant failed QC.</small></p>

<small>Hovering over a failing genotype will list the filter applied. Genotypes are shown as follows:</small>
<div class='panel'>
<div class="panel panel-default">
  <div class="panel-body">
    <span class="label gt-0 PASS">Passing REF</span>&nbsp;
    <span class="label gt-2 PASS">Passing ALT</span>&nbsp;
    <span class="label gt-0 mapping_quality min_depth dv_dp het">Failing REF</span>&nbsp;
    <span class="label gt-2 mapping_quality min_depth dv_dp het">Failing ALT</span>&nbsp;
    <span class="label gt-1 het">Heterozygous</span>
  </div>
</div>
</div>

<table style='border-radius: 5px' class="table table-hover table-striped table-condensed table-bordered table-responsive" id="variants">
    <thead>
      <tr>
        <th>CHROM:POS</th>
        <th>REF / ALT</th>
        <th>Filter</th>
        <th><span class='tooltip-item' data-toggle='tooltip' title='Allele Frequency'>AF</span></th>
        <th>Gene Name</th>
        <th>WormBase ID</th>
        <th>Sequence</th>
        <th>Coding Change</th>
        <th>Biotype</th>
        <th>Annotation</th>
        <th>Impact</th>
      </tr>
    </thead>
    <tbody></tbody>
</table>


      </div>
  </div>
{% endblock %}

{% block script %}


<script type="text/javascript">

trackset = {"Genes": {
                            name: "Genes",
                            displayMode: "EXPANDED",
                            order: 1,
                            url: "//storage.googleapis.com/elegansvariation.org/releases/{{ DATASET_RELEASE }}/browser/elegans_gene.WS245.bed",
                            indexed: false,
                            searchable: false,
                            color: "#5c5cd6",
                            height: 30
                     },
            "Transcripts": {
                            name: "Transcripts",
                            url: "//storage.googleapis.com/elegansvariation.org/releases/{{ DATASET_RELEASE }}/browser/elegans_transcripts_WS245.bed",
                            order: 2,
                            color: "#a366ff",
                            displayMode: "SQUISHED",
                            searchable: false
                           },
            "LOW": {
                            name: "LOW",
                            url: "//storage.googleapis.com/elegansvariation.org/releases/{{ DATASET_RELEASE }}/tracks/severity/{{ DATASET_RELEASE }}.LOW.bed",
                            order: 3,
                            color: "#33cc33",
                            displayMode: "EXPANDED",
                            height: 20
                   },
            "MODERATE": {
                            name: "MODERATE",
                            url: "//storage.googleapis.com/elegansvariation.org/releases/{{ DATASET_RELEASE }}/tracks/severity/{{ DATASET_RELEASE }}.MODERATE.bed",
                            order: 4,
                            color: "#ffc500",
                            displayMode: "EXPANDED",
                            height: 20
                   },
            "HIGH": {
                            name: "HIGH",
                            url: "//storage.googleapis.com/elegansvariation.org/releases/{{ DATASET_RELEASE }}/tracks/severity/{{ DATASET_RELEASE }}.HIGH.bed",
                            order: 5,
                            color: "#ff0000",
                            displayMode: "EXPANDED",
                            height: 20
                   },
            "MODIFIER": {
                            name: "MODIFIER",
                            url: "//storage.googleapis.com/elegansvariation.org/releases/{{ DATASET_RELEASE }}/tracks/severity/{{ DATASET_RELEASE }}.MODIFIER.bed",
                            order: 5,
                            color: "#999999",
                            displayMode: "EXPANDED",
                            height: 20
                   },
            "Variants": {
                            name: "Variants",
                            url: "//storage.googleapis.com/elegansvariation.org/releases/{{ DATASET_RELEASE }}/WI.{{ DATASET_RELEASE }}.vcf.gz",
                            indexURL: "//storage.googleapis.com/elegansvariation.org/releases/{{ DATASET_RELEASE }}/WI.{{ DATASET_RELEASE }}.vcf.gz.tbi",
                            order: Number.MAX_VALUE,
                            displayMode: "SQUISHED",
                            color: "#000000",
                            homvarColor: "#0066ff",
                            homrefColor: "#c2c2d6",
                            visibilityWindow: 20000
                        },
            "phastCons": {
                name: "phastCons",
                url: "//storage.googleapis.com/elegansvariation.org/browser_tracks/elegans.phastcons.bw",
                order: 6,
                displayMode: "SQUISHED",
                color: "#000000",
                visibilityWindow: 20000
            },
            "phyloP": {
                name: "phylop",
                url: "//storage.googleapis.com/elegansvariation.org/browser_tracks/elegans.phylop.bw",
                order: 6,
                displayMode: "SQUISHED",
                color: "#000000",
                visibilityWindow: 20000
            },

            {% for isotype in isotype_listing %}
            {{ isotype }} : {
              name: "{{ isotype }}",
              url: "//storage.googleapis.com/elegansvariation.org/releases/{{ DATASET_RELEASE }}/tracks/isotype/{{ isotype }}.{{ DATASET_RELEASE }}.vcf.gz",
              order: 100,
              displayMode: "EXPANDED",
              color: "#ffffff",
              homvarColor: "#0066ff",
              homrefColor: "#c2c2d6",
              visibilityWindow: 20000,
              searchable: false
            },
            {{ isotype }}_bam : {
              id: "{{ isotype }}_bam",
              name: "{{ isotype }}",
              url: "//s3.us-east-2.amazonaws.com/elegansvariation.org/bam/{{ isotype }}.bam",
              order: 100,
              visibilityWindow: 20000,
              searchable: false
            },
            {% endfor %}



}

$(document).ready(function () {
      {% for k,v in impact_items %}
        {% if v in var_eff %}
          $('.track-select[value="{{ v }}"]').prop('checked', true);
        {% endif %}
      {% endfor %}

        var div = $("#browser")[0],
                options = {
                    search: {
                      url: "/api/browser/search/$FEATURE$",
                      coords: 1,
                      resultsField: 'result'
                    },
                    showNavigation: true,
                    showKaryo: false,
                    reference: {
                      id: "WS245",
                      fastaURL: "//storage.googleapis.com/andersen_dist/genome/c_elegans.PRJNA13758.WS245.genomic.fa",
                    },
                    locus: "{{ region }}",
                    tracks: [
                        {% for i in var_eff %}
                          trackset["{{ i }}"],
                        {% endfor %}
                        trackset["Genes"],
                        trackset["Transcripts"],
                        trackset["MODERATE"],
                        trackset["HIGH"]
                        ],
                };
        igv.createBrowser(div, options);

// Detect track changes
$(".track-select").on("change", function() { 
  if ($(this).prop("checked") == true) {
    igv.browser.loadTrack(trackset[$(this).attr("value")]);
  } else {
    remove_track = $(this).attr("value");
    $.each(igv.browser.trackViews, function(k, v) {
      if(v.track.id == remove_track) {
        igv.browser.removeTrack(v.track);
      }
    })
  }
})

var_impact_class = {"LOW": "success",
                    "MODERATE":"warning",
                    "HIGH": "danger",
                    "MODIFIER": "default"}


function process_gene_search() {
    $("#loading-search-table").fadeOut();
    var gene = $('#gene-search').val();
    if (gene.length == 0) {
      $("#search-table").fadeOut();
    } else {
      $("#orthologs").html("");
      $.ajax({
           url: "/api/gene/browser-search/" + gene,
           method: "GET",
           contentType: 'application/xml',
           }).done(function(msg) {
            row = Array();
            $.each(msg, function(i,row) {
                if ("chrom" in row) {
                   link = row["chrom"] + ":" + row["start"] + "-" + row["end"];
                } else {
                    link = row["locus"];
                }
                gene_name = `<a link='${link}' class='ortholink'>${row["locus"]}</a>`;
                homolog_species = row["homolog_species"] || "<em>C. elegans</em>";
                homolog_source = row["homolog_source"] || "Wormbase";
                result = [gene_name, row['homolog_gene'] || row["gene_symbol"], homolog_species, row["homolog_source"] || "Wormbase"]  ;
                result = "<tr><td>" + result.join("</td><td>") + "</td></tr>";
                position = row["chrom"] + ":" + row["start"];
              $("#orthologs").append(result);
            });
            $("#search-table").fadeIn();
            });
    }
}

// Make links work!  
$(".container-fluid").on("click", ".ortholink", function() {
  igv.browser.search($(this).attr("link"))
  // Update tables and URL
  refresh_variants();
});


function gt_label(gt) {
  // Generates a genotype label
  r = ""
  classes = [];
  classes.push(`gt-${gt["GT"]}`);
  FT = gt["FT"].split(";").join(" ");
  classes.push(FT);
  tt = "";
  if (gt['FT'] != ['PASS']) {
      tt = ` data-placement='bottom' title='${FT}' `;
  }
  r += `<div class='label ttop ${classes.join(" ")}' ${tt} >`;
  r += gt["SAMPLE"] + " : " + gt["TGT"];
  r += "</div>";
  return r
}

function draw_gt_set(genotype_set, genotype_val) {
    return genotype_set.filter(function(gt) { return gt['GT'] == genotype_val })
                       .map( gt_label )
                       .join(" ");

}

var xhr;
function refresh_variants(download = false) {
    region = $(".igvNavigationSearchInput").val();
    if (/[IVXMtDNA]+:[0-9,]+-[0-9,]+/g.exec(region) != null) {
      chrom = region.split(":")[0];
      start = region.split(":")[1].split("-")[0].replace(/,/g, "");
      end = region.split(":")[1].split("-")[1].replace(/,/g, "");

      sample_tracks = []
      $(".sample-track:checked").each(function() { sample_tracks.push(this.value) })

      sample_tracks_alignment = []
      $(".sample-track-alignment:checked").each(function() { sample_tracks_alignment.push(this.value) })

      normal_tracks = []
      $(".normal-track:checked").each(function() { normal_tracks.push(this.value) })
      variant_impact = []
      $(".variant-impact:checked").each(function() { variant_impact.push(this.value) })

      track_set = $.extend(normal_tracks, variant_impact, sample_tracks, sample_tracks_alignment);
      tracks = [];
      for (track in track_set) {
           if ($(`.track-select[value=${track}]`).prop("checked") == true){
              tracks += track_set[track];
           }
      }
      
      data = {'chrom': chrom,
              'start': parseInt(start),
              'end': parseInt(end),
              'variant_impact': variant_impact.join('_'),
              'sample_tracks': sample_tracks.join('_'),
              'list-all-strains': $('.list-all-strains').prop('checked')
             }

      if (download == true) {
          data['output'] = 'tsv';
          url =  "{{ url_for('variant_query') }}?" + $.param(data);
          window.open(url, sprintf("%s_%s_%s.csv", chrom, start, end));
      } else {

        $("#variants > tbody").html("");
        data['output'] = 'json';
        data =  $.param(data);

        if(xhr) {
          xhr.abort();
        }

        xhr = $.ajax({
          url: "{{ url_for('variant_query') }}",
          data: data,
          method: "GET",
          contentType: "application/json",
          success: function(msg) {
              
              $.each(msg, function(k, rec) {
                var td_rowspan = 1;
                if (variant_impact.indexOf("ALL") > -1 && rec["ANN"].length == 0) {
                  rec["ANN"] = [{ effect: "", gene_id: null}]
                }
                var is_first = true;
                $.each(rec['ANN'], function(annk, ann) {

                        if (ann["gene_id"] != null) 
                        {
                          ann["gene_id"] = `<a href='/gene/${ann["gene_id"]}'>${ann["gene_id"]}</a>`;
                        }
                        if (ann.impact != undefined) {
                          impact = `<span style='color: white;' class='label label-${var_impact_class[ann.impact]}'>${ann.impact}</span>`
                        } else {
                          impact = ""
                        }

                        items = [ann.gene_name,
                                 ann.gene_id,
                                 ann.feature_id,
                                 ann.aa_change,
                                 ann.transcript_biotype,
                                 ann.effect.split("&").join(", "),
                                 impact];

                        if (is_first) {
                            td_rowspan = rec['ANN'].length;
                            variant_items = [`${rec["CHROM"]}:${Number(rec["POS"]).toLocaleString('en')}`,
                                             `${rec["REF"]} / ${rec["ALT"].join(",")}`,
                                             rec["FILTER"],
                                             rec["AF"]]
                            is_first = false;
                            rowspan_td = `<td class='bold' rowspan=${td_rowspan}>`;
                            variant_items = rowspan_td + variant_items.join(`</td>${rowspan_td}`) + "</td>"
                        } else {
                            variant_items = ""
                        }

                        additional_fields = [ann.gene_name + ": " , ann.filter]
                        if (rec['FILTER'] != "PASS") {
                          fail_variant_class = 'class="warning"'
                        } else {
                          fail_variant_class = ""
                        }
                        $("#variants > tbody").append(`<tr ${fail_variant_class}>${variant_items}<td>${items.join("</td><td>")}</td></tr>`);
                    }
                );

                if ($(".list-all-strains").prop('checked') == true) {
                    gt_panel = $(".isotype-item").map(
                    function () {return this.value;}).get();
                } else {
                    gt_panel = $(".isotype-item:checked").map(
                    function () {return this.value;}).get();
                }
                
                if (gt_panel.length > 0) {
                  gts = rec["GT"].filter(function(i) { return gt_panel.includes(i["SAMPLE"]) });
                    // Append genotypes
                    $("#variants > tbody").append(
                      `<tr><td style='border-top:none;' colspan='11'>
                        <div class='row'>
                          <div class='col-md-12 gt_set'>
                              <h5>Reference</h5>
                              ${draw_gt_set(gts, 0)}
                              <h5>Alternative</h5>
                              ${draw_gt_set(gts, 2)}
                              <h5>Other</h5>
                              ${draw_gt_set(gts, 1)}
                              ${draw_gt_set(gts, 3)}
                          </div>
                        </div>
                      </td></tr>`);
                    // enable tooltips
                    $('.ttop').tooltip();
                }

            });


          


            // Enable links
            $( "#variants a" ).on( "click", function() {
                window.location = $(this).attr('href');
            });

          }
        });

        } // End if else logic for download
        } // Test for region

}

$("#download").click(function() {
    refresh_variants(download=true);
});


$("#browser").on("mouseup", function() { 
  setTimeout(refresh_variants, 200); 
})


var typingTimer;                //timer identifier
var doneTypingInterval = 1000;  //time in ms (5 seconds)

// Ortholog search
$("#gene-search").on("input", function(e) {
  $("#loading-search-table").fadeIn();
  clearTimeout(typingTimer);
  typingTimer = setTimeout(process_gene_search, doneTypingInterval);
})

$(".checkbox").on("click", function() { 
  setTimeout(refresh_variants, 200); 
})

$(".igvNavigationSearchInput").on("input paste", function() {
  refresh_variants();
});

$(".track-select").on("change", function() {
  refresh_variants();
});

$(".igvNavigationSearchInput").keypress(function(e) {
    if(e.which == 13) {
      setTimeout(refresh_variants, 200); 
    }
});



// Initial load
setTimeout(refresh_variants, 1000);

});



$(document).ready(function() {

    (function($) {
        var patterns = [];
        $('#filter').keyup(function() {
            $('.searchable tr').hide();
            $(this).val().split(',').forEach(function(r) {
                var rex = new RegExp(r, "i");
                $('.searchable tr').filter(function() {
                    return rex.test($(this).text());
                }).show();
            })  
        })

    }(jQuery));


    $('#filter').keydown(function(event) {
        if (event.keyCode == 13) {
            event.preventDefault();
            return false;
        }
    });

});





</script>



{% endblock %}
