{% extends "_layouts/default.html" %}

{% block custom_head %}

<!-- jQuery UI CSS -->
<link rel="stylesheet" type="text/css"
      href="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/themes/smoothness/jquery-ui.css"/>

<!-- Font Awesome CSS -->
<link rel="stylesheet" type="text/css"
      href="//maxcdn.bootstrapcdn.com/font-awesome/4.2.0/css/font-awesome.min.css"/>

<!-- IGV CSS -->
<link rel="stylesheet" type="text/css" href="//igv.org/web/release/1.0.0-rc1/igv-1.0.0-rc1.css">

<!-- jQuery JS -->
<script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/jquery-ui.min.js"></script>


<script type="text/javascript" src="{{ url_for('static', filename = 'js/igv-rc1.js') }}"></script>

<style>

.gt-00 {
  background-color: rgba(194,194,214,1.0);
  border: 1px solid black;
  color: black;
}
.gt-11 {
  background-color: rgba(0, 102, 255,1.0);
  border: 1px solid black;
  color: white;
}

.gt-00 .mapping_quality,.min_depth,.dv_dp,.het {
  background-color: rgba(194,194,214,0.25);
  border: none;
}
.gt-11 .mapping_quality,.min_depth,.dv_dp,.het{
  background-color: rgba(0, 102, 255,0.25);
  border: none;
}

</style>

{% endblock %}


{% block content %}
<div class="row">
      <div class="col-md-12">
    <div class="well">
    <div class="row">
      <div class="col-md-2">
      <h5>Tracks <a href="{{ url_for('help_item', filename = 'Variant-Prediction') }}#standard-tracks"><span class="glyphicon glyphicon-question-sign"></span></a></h5>
      {% for i in ["Genes", "Transcripts"] %}
        <div class="checkbox">
          <label><input type="checkbox" checked class="track-select" value="{{ i }}" />{{ i }}</label>
        </div>
      {% endfor %}
      {% for i in ["phyloP", "phastCons", "Variants"] %}
        <div class="checkbox">
          <label><input type="checkbox" class="track-select" value="{{ i }}" />{{ i }}</label> 
        </div>
      {% endfor %}

        <h5>Variant Effects <a href="{{ url_for('help_item', filename = 'Variant-Browser') }}#variant-effects"><span class="glyphicon glyphicon-question-sign"></span></a></h5>

        <div class="checkbox">
            <label><input type="checkbox" class="track-select" value="LOW" />LOW <div class="legend-box" style="background-color: #66d866;"></div></label>
        </div>

        <div class="checkbox">
              <label><input type="checkbox" class="track-select" value="MODERATE" />MODERATE <div class="legend-box" style="background-color: #ffd33f;"></div></label>
        </div>

        <div class="checkbox">
              <label><input type="checkbox" class="track-select" value="HIGH" />HIGH <div class="legend-box" style="background-color: #ff3f3f;"></div></label>
        </div>


      </div>
      <div class="col-md-6">

      <h5>Gene Search</h5>
        <input type="text" class="form-control" id="gene-search" placeholder="Gene Search (e.g. trt-1)">
      <div id="loading-search-table" style="display:none; margin-top:20px;"><div class="spinner"><div class="bounce1"></div><div class="bounce2"></div><div class="bounce3"></div></div></div>
      
      <table id="search-table" class='table table-striped' style="display: none;">
      <thead>
      <tr>
        <th><em>C. elegans Gene</em></th>
        <th>Symbol/ID</th>
        <th>Sequence</th>
        <th>Species</th>
        <th>Type</th>
      </tr>
      </thead>
    <tbody id="orthologs"></tbody>
    </table>

      </div>

      <div class="col-md-4">
        <h5>Strains <a href="{{ url_for('help_item', filename = 'Variant-Browser') }}#ind-strains"><span class="glyphicon glyphicon-question-sign"></span></a></h5>

        <form class="form-inline pull-xs-right">

          <input class="form-control" id="filter" type="text" style="width:100%;  margin-bottom: 5px;" placeholder="e.g. JU360" autocomplete="off">

        </form>

        <table id='browser_strain_list' class="table table-striped table-hover table">
          
          <tbody class="searchable">

            {% for isotype in isotypes %}

              <tr class="isotype-select">

              <td class='isotype-item'><label><input type="checkbox" class="track-select" value="{{ isotype }}" /> {{ isotype }}</label></td>

              </tr>

            {% endfor %}

          </tbody>

        </table>

      </div>

    </div>
     
    </div>
</div>{# Close col #}
</div>{# Close row #}


<div id="browser"></div>

<div id="row" id="browser-legend">

    <div class="col-lg-12">
        <div class="legend-box" style="background-color: #c2c2d6;"></div> <strong>Reference</strong>&nbsp;&nbsp;
        <div class="legend-box" style="background-color: #0066ff;"></div> <strong>Alternate</strong>
          <label><div class="legend-box" style="background-color: #66d866;"> </div> LOW</label>&nbsp;&nbsp;
        <label><div class="legend-box" style="background-color: #ffd33f;"></div> MODERATE</label>
        <label><div class="legend-box" style="background-color: #ff3f3f;"></div> HIGH</label>
    </div>

</div>


<div class="row">
<div class="col-md-12 col-lg-12">
<h3>Variants</h3>

<table class="table table-striped table-hover" id="variants">

<p><small>Maximum of 1000 variants from the region will be listed. Variants without a black border have a filter applied, and are less reliable.</small></p>
  
<thead>
  <tr>
    <th>CHROM:POS</th>
    <th>REF / ALT</th>
    <th>Gene Name</th>
    <th>WormBase ID</th>
    <th>Sequence</th>
    <th>Coding Change</th>
    <th>Biotype</th>
    <th>Annotation</th>
    <th>Impact</th>
  </tr>
</thead>

<tbody>
</tbody>

</table>
      
      </div>
  </div>
{% endblock %}

{% block script %}


<script type="text/javascript">

trackset = {"Genes": {
                            name: "Genes",
                            displayMode: "EXPANDED",
                            order: 1,
                            url: "https://storage.googleapis.com/andersen_dist/vcf/all/{{ build }}/browser/elegans_gene.WS245.bed",
                            indexed: false,
                            searchable: true,
                            color: "#5c5cd6",
                            height: 30
                     },
            "Transcripts": {
                            name: "Transcripts",
                            url: "https://storage.googleapis.com/andersen_dist/vcf/all/{{ build }}/browser/elegans_transcripts_WS245.bed",
                            indexURL: "https://storage.googleapis.com/andersen_dist/vcf/all/{{ build }}/browser/elegans_transcripts_WS245.bed.idx",
                            order: 2,
                            color: "#a366ff",
                            displayMode: "SQUISHED",
                            searchable: true
                           },
            "LOW": {
                            name: "LOW",
                            url: "https://storage.googleapis.com/andersen_dist/vcf/all/{{ build }}/browser/{{ build }}.LOW.bed",
                            indexURL: "https://storage.googleapis.com/andersen_dist/vcf/all/{{ build }}/browser/{{ build }}.LOW.bed.idx",
                            order: 3,
                            color: "#33cc33",
                            displayMode: "EXPANDED",
                            height: 20
                   },
            "MODERATE": {
                            name: "MODERATE",
                            url: "https://storage.googleapis.com/andersen_dist/vcf/all/{{ build }}/browser/{{ build }}.MODERATE.bed",
                            indexURL: "https://storage.googleapis.com/andersen_dist/vcf/all/{{ build }}/browser/{{ build }}.MODERATE.bed.idx",
                            order: 4,
                            color: "#ffc500",
                            displayMode: "EXPANDED",
                            height: 20
                   },
            "HIGH": {
                            name: "HIGH",
                            sourceType: "gcs",
                            url: "https://storage.googleapis.com/andersen_dist/vcf/all/{{ build }}/browser/{{ build }}.HIGH.bed",
                            indexURL: "https://storage.googleapis.com/andersen_dist/vcf/all/{{ build }}/browser/{{ build }}.HIGH.bed.idx",
                            order: 5,
                            color: "#ff0000",
                            displayMode: "EXPANDED",
                            height: 20
                   },
            "Variants": {
                            name: "Variants",
                            url: "//storage.googleapis.com/andersen_dist/vcf/all/{{ build }}/WI.{{ build }}.filtered.vcf.gz",
                            indexURL: "//storage.googleapis.com/andersen_dist/vcf/all/{{ build }}/WI.{{ build }}.filtered.vcf.gz.tbi",
                            order: Number.MAX_VALUE,
                            displayMode: "SQUISHED",
                            color: "#000000",
                            homvarColor: "#0066ff",
                            homrefColor: "#c2c2d6",
                            visibilityWindow: 20000
                        },
            "phastCons": {
                name: "phastCons",
                url: "//storage.googleapis.com/andersen_dist/genome/elegans.phastcons.bw",
                order: 6,
                displayMode: "SQUISHED",
                color: "#000000",
                visibilityWindow: 20000
            },
            "phyloP": {
                name: "phylop",
                url: "//storage.googleapis.com/andersen_dist/genome/elegans.phylop.bw",
                order: 6,
                displayMode: "SQUISHED",
                color: "#000000",
                visibilityWindow: 20000
            },

            {% for strain in isotypes %}
            {{ strain }} : {
              name: "{{ strain }}",
              url: "//storage.googleapis.com/andersen_dist/vcf/all/{{ build }}/browser/{{ strain }}.vcf.gz",
              order: 100,
              displayMode: "EXPANDED",
              color: "#ffffff",
              homvarColor: "#0066ff",
              homrefColor: "#c2c2d6",
              visibilityWindow: 20000,
              searchable: false
            },
            {% endfor %}



}

$(document).ready(function () {
      {% for k,v in impact_items %}
        {% if v in var_eff %}
          $('.track-select[value="{{ v }}"]').prop('checked',true);
        {% endif %}
      {% endfor %}

        var div = $("#browser")[0],
                options = {
                    showNavigation: true,
                    showKaryo: false,
                    reference: {
                      id: "WS245",
                      fastaURL: "//storage.googleapis.com/andersen_dist/genome/c_elegans.PRJNA13758.WS245.genomic.fa",
                      indexURL: "//storage.googleapis.com/andersen_dist/genome/c_elegans.PRJNA13758.WS245.genomic.fa.fai",
                      // cytobandURL: ""
                    },
                    locus: "{{ region }}",
                    tracks: [
                        {% for i in var_eff %}
                          trackset["{{ i }}"],
                        {% endfor %}
                        trackset["Genes"],
                        trackset["Transcripts"]                    ],
                };
        igv.createBrowser(div, options);

// Detect track changes
$(".track-select").on("change", function() { 
  if ($(this).prop("checked") == true) {
    igv.browser.loadTrack(trackset[$(this).attr("value")]);
  } else {
    remove_track = $(this).attr("value");
    $.each(igv.browser.trackViews, function(k, v) {
      if(v.track.name == remove_track) {
        igv.browser.removeTrack(v.track);
      }
    })
  }
})

var_impact_class = {"LOW": "success",
                    "MODERATE":"warning",
                    "HIGH": "danger",
                    "MODIFIER": "default"}


function process_gene_search() {
    $("#loading-search-table").fadeOut();
    var gene = $('#gene-search').val();
    if (gene.length == 0) {
      $("#search-table").fadeOut();
    } else {
      $("#orthologs").html("");
      $.ajax({
           url: "/api/gene/search/" + gene,
           method: "GET",
           contentType: 'application/xml',
           }).done(function(msg) {
            row = Array();
            $.each(msg, function(i,x) {
              if ("CHROM" in x) {
                 link = x["CHROM"] + ":" + x["start"] + "-" + x["end"];
              } else {
                  link = x["ce_gene_name"];
              }
              gene_name = "<a link='" + link + "' class='ortholink'>" + x["ce_gene_name"] + "</a>";
              row = [gene_name, x["gene_symbol"], x["species"], x["source"]];
              row = "<tr><td>" + row.join("</td><td>") + "</td></tr>";
              $("#orthologs").append(row);
            });
            $("#search-table").fadeIn();

            // Make links work!  
            $(".ortholink").on("click", function() {
              igv.browser.search($(this).attr("link"))
              // Update tables and URL
              refresh_variants();
            });
            });
    }
}

function encodeQueryData(data) {
   let ret = [];
   for (let d in data)
     ret.push(encodeURIComponent(d) + '=' + encodeURIComponent(data[d]));
   return ret.join('&');
}

function refresh_variants() {
    region = $(".igvNavigationSearchInput").val();
    chrom = region.split(":")[0];
    start = region.split(":")[1].split("-")[0].replace(/,/g, "");
    end = region.split(":")[1].split("-")[1].replace(/,/g, "");
    track_set = {'LOW':'l', 'MODERATE':'m', 'HIGH':'h'};
    tracks = "";
    for (track in track_set) {
         if ($(".track-select[value="+track+"]").prop("checked") == true){
            tracks += track_set[track];
         }
    }

    $("#variants > tbody").html("");
    
    data = {'output_all_variants': false,
            'trackset': track_set}
    var querystring = encodeQueryData(data);
    var loc = chrom + ":" + start + "-" + end + "/" + tracks + "_?" + querystring;
    // update browser

    window.history.pushState({"pageTitle": document.title + "|" + loc},"", "/data/browser/" + loc);

    $.getJSON("/api/variant/" + loc, function(msg) {
          $.each(msg, function(k, r) {
            $.each(r['ANN'], function(annk, a) {
              console.log(a);

            if (a["gene_id"] != "") 
            {
              a["gene_id"] = "<a href='/gene/" + a["gene_id"] + "'>" + a["gene_id"] + "</a>";
            }

            items = [r["CHROM"] + ":" + Number(r["POS"]).toLocaleString('en'),
                     r["REF"] + " / " + r["ALT"].join(","),
                     a["gene_name"],
                     a["gene_id"],
                     a["feature_id"],
                     a["hgvs_p"],
                     a["transcript_biotype"],
                     a["effect"].split("&").join(", "),
                     "<span class='label label-" + var_impact_class[a["impact"]] + "'>" + a["impact"] + "</span>"];

            //suppose to happen on click
            additional_fields = [a["gene_name"] + ": " , a["filter"]]
            $("#variants > tbody").append("<tr><td>" + items.join("</td><td>") + "</td></tr>");
            checked_isotypes = $(".isotype-item input:checked").map(
                function () {return this.value;}).get();
            
            if (checked_isotypes.length > 0) {
              gts = a["GT"].filter(function(i) { return checked_isotypes.includes(i["SAMPLE"]) });

              // Construct gt view
              gtset = gts.map( function(gt) {

                r = ""
                classes = [];
                classes.push("gt-" + gt["GT"].replace("/", ""));
                classes.push(gt["FT"].split(";"));
                r += "<span class='label " + classes.join(" ") + "'>";
                r += gt["SAMPLE"] + " : " + gt["TGT"];
                r += "</span>";
                return r}).join(" ");

              $("#variants > tbody").append("<tr class='" + var_impact_class[a["impact"]] + "'><td style='border-top:none;' colspan='9'>" + gtset + "</td>" + "</tr>");
            }
          });
        });


        


          // Enable links
          $( "#variants a" ).on( "click", function() {
              window.location = $(this).attr('href');
          });

        });


}


$("#browser").on("mouseup", function() { 
  setTimeout(refresh_variants, 200); 
})


var typingTimer;                //timer identifier
var doneTypingInterval = 1000;  //time in ms (5 seconds)

// Ortholog search
$("#gene-search").on("input", function(e) {
  $("#loading-search-table").fadeIn();
  clearTimeout(typingTimer);
  typingTimer = setTimeout(process_gene_search, doneTypingInterval);
})

$(".checkbox").on("click", function() { 
  setTimeout(refresh_variants, 200); 
})

$(".igvNavigationSearchInput").on("input paste", function() {
  refresh_variants();
});

$(".track-select").on("change", function() {
  refresh_variants();
});

$(".igvNavigationSearchInput").keypress(function(e) {
    if(e.which == 13) {
      setTimeout(refresh_variants, 200); 
    }
});



// Initial load
setTimeout(refresh_variants, 1000);

});


</script>


<script type="text/javascript">
$(document).ready(function() {

    (function($) {
        var patterns = [];
        $('#filter').keyup(function() {
            $('.searchable tr').hide();
            $(this).val().split(',').forEach(function(r) {
                var rex = new RegExp(r, "i");
                $('.searchable tr').filter(function() {
                    return rex.test($(this).text());
                }).show();
            })  
        })

    }(jQuery));


    $('#filter').keydown(function(event) {
        if (event.keyCode == 13) {
            event.preventDefault();
            return false;
        }
    });

});

</script>



{% endblock %}
