{% extends "_layouts/default.html" %}


{% block custom_head %}
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<script src="https://cdn.plot.ly/plotly-latest.min.js"></script>

<link rel="stylesheet" href="/static/css/leaflet.css" />
<script src="/static/js/leaflet.js"></script>
<script src="/static/js/leaflet-search.js"></script>



<script type="text/x-mathjax-config">
  MathJax.Hub.Config({tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}});
</script>
<script type="text/javascript" async
  src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_CHTML">
</script>


<style>

.report {
  width: 100%;
}

.height-report {
  height: 60%;
  width: 60%;
  display: block;
  margin-left: auto;
  margin-right: auto
}

#report_nav.affix {
    position: fixed;
    top: 0px;
    left: 0px;
    width: 100%;
}

</style>


{% endblock %}


{% block style %}
{% endblock %}

{% block content %}

<div class="row">
      <div class="col-md-12">

      <div class="btn-group" role="group">
        {% for trait in report_data %} 

          <a href="{{ url_for('trait_view', report_slug = report_url_slug, trait_slug = trait.trait_slug) }}" type="button" class="btnbtn-secondary btn-sm {% if trait.trait_slug == trait_slug %} btn-primary {% endif %}">{{ trait.trait_name }}</a>     
          
        {% endfor %}

        </div>
        <hr />
      </div>
</div>

{% if trait_data.status == "complete" %}

<div class="row">

<div class="col-md-12">

<div class="panel panel-default">
  <div class="panel-heading">
    <h3 class="panel-title">{{ trait_data["trait_name"] }}</h3>
  </div>
  <div class="panel-body">
  <div class="row">
  <div class='col-md-6'>
  <strong>Data</strong>
    <ul>
      {% if 'raw_phenotype.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/raw_phenotype.tsv">Raw Phenotype Data</a></strong></li>
      {% endif %}

      {% if 'processed_phenotype.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/processed_phenotype.tsv">Processed Phenotype Data</a></strong></li>
      {% endif %}

      {% if 'interval_variants.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/interval_variants.tsv">Interval Variants</a></strong></li>
      {% endif %}

      {% if 'raw_mapping.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/raw_mapping.tsv">Raw Mappings</a></strong></li>
      {% endif %}

      {% if 'processed_sig_mapping.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/processed_sig_mapping.tsv">Processed Significant Mappings</a></strong></li>
      {% endif %}

      {% if 'non_sig_mapping.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/non_sig_mapping.tsv">Nonsignificant Mappings</a></strong></li>
      {% endif %}
    </ul>
    </div>
    <div class='col-md-6'>
    <strong>Info</strong>
    <ul>
      <li><strong>Version</strong>: {{ trait_data.version }}</li>
      <li><strong>Availibility</strong>: {% if trait_data.release == 0 %}
                                            Public
                                         {% elif trait_data.release == 1 %}
                                            Embargo
                                         {% else %}
                                            Private
                                         {% endif %}</li>
    </ul>
    </div>
    </div>
  </div>
</div>

  </div>
</div>


<div id="report" data-toggle="tab">


<nav id="report_nav" data-offset-top="385" data-spy="affix" class="navbar navbar-default" style="z-index:10000">
  <div class="container-fluid">
   <!-- Collect the nav links, forms, and other content for toggling -->
    <div class="collapse navbar-collapse">
      <ul class="nav navbar-nav">
        <li><a href="#phenotype">Phenotype</a></li>
        <li><a href="#manhattan">Manhattan Plot</a></li>
        <li><a href="#intervals">Intervals</a></li>
        <li><a href="#pxg">PxG</a></li>
        <li><a href="#ld">LD</a></li>
        <li><a href="#peak_summary">Peak Summary</a></li>
      </ul>
    </div><!-- /.navbar-collapse -->
  </div><!-- /.container-fluid -->
</nav>


<h2 id="phenotype" name="phenotype">Phenotype</h2>


<img class="report" src="{{ base_url }}/figures/phenotype_histogram-1.png" />
<strong>Histogram representation of your phenotype</strong>



  Your phenotype is significantly associated with genetic variation present in the <em>C. elegans</em> population!

<!--
  Unfortunately your phenotype was not significantly associated with genetic variation present in the <em>C. elegans</em> population. This could be due to noisy trait data - have you performed heritability analysis for your phenotype using our heritability strain panel? If you have and the heritability was found to be high for this trait, the trait might not have reached statistical significance because it is highly complex and more strains need to be phenotyped. If you have phenotyped the entire 152 wild-isolate collection, you can patiently wait for more isolates to be added to the collection, or generate an F2 recombinant inbred line (RIL) panel generated between strains with high and low phenotypes.
 -->



<a name="manhattan"></a>
<h2 id="manhattan">Manhattan Plot</h2>

<img class="report" src="{{ base_url }}/figures/Manplot-1.png" />

A genome-wide representation of the association between variation in the <em>C. elegans</em>em> population and your phenotype. The x-axis corresponds to genomic position with chromosome number indicated in the gray box above each facet. Dots represent single-nucleotide variant (SNV) present in the <em>C. elegans</em> population. The y-axis corresponds to the level of significance for the association test. Blue dots represent SNVs that are above the defined significance threshold, which is the thick gray line. Red boxes surrounding blue dots represent the QTL confidence interval, which we define as plus/minus 50 SNVs from the last signifincant SNV. The default threshold for significance is the Bonferroni-corrected value $(-log_{10}\frac{0.05}{\#SNVs})$ and is usually around 5.5.

<h3 id="intervals">QTL Confidence Intervals</h3>

<table class="table table-hover table-striped confidence_intervals">
<thead>
  <tr>
    <th>Interval</th>
    <th>Peak Position</th>
    <th>-log<sub>10</sub>p</th>
    <th>Genes</th>
    <th>Protein Coding</th>
  </tr>
</thead>
<tbody>
{% for row in mapping_results %}
<tr chrom="{{ row.chrom }}" start="{{ row.interval_start }}" end="{{ row.interval_end }}">
  <td><a href="http://www.wormbase.org/tools/genome/gbrowse/c_elegans_PRJNA13758/?name={{ row.chrom}}:{{ row.interval_start }}..{{ row.interval_end }}" target="_blank">{{ row.chrom }}:{{ "{:,.0f}".format(row.interval_start) }}-{{ "{:,.0f}".format(row.interval_end) }}</a></td>
  <td><a href="http://www.wormbase.org/tools/genome/gbrowse/c_elegans_PRJNA13758/?name={{ row.chrom}}:{{ row.interval_start }}..{{ row.interval_end }}" target="_blank">{{row.chrom}}:{{ "{:,.0f}".format(row.pos) }}</a></td>
  <td>{{ row.log10p|round(3) }}</td>
  <td>-</td>
  <td>-</td>
</tr>
{% endfor %}
</tbody>
</table>
<h3 id="pxg">Phenotype-by-Genotype</h3>
<br />
<img class="report" src="{{ base_url }}/figures/PxGplot-1.png" />

    
<!--<div id="chartContainer"></div>-->


<p>Phenotypic distributions represeted as box plots are split by the genotype at the most significant SNV for a given QTL. The phenotype is on the y-axis and the genotype is on the x-axis. The chromosome and position for the plotted SNV are indicated in each the gray rectangles above the panels. REF refers to the N2 genotype and ALT refers to the variant phenotype. All of the SNVs in our data set are biallelic so these are the only two classes for any given SNV.</p>

<h3 id="ld">Linkage Disequilibrium</h3>

<img class="height-report" src="{{ base_url }}/figures/LDplot-1.png" />

<p>Peak markers are the most statistically significant markers identified within a peak. The Linkage Disequilibrium (LD) measure being plotted is the correlation between peak markers $r = \frac{-D}{\sqrt{p(A) * p(a) * p(B) * p*(b)}}$, where D raw difference in frequency between the observed number of AB pairs and the expected number and A, B, a, and b refer to alleles at the loci. An LD value of 0.8 or higher suggests that the two peak markers are not segregating randomly.</p>
<br />
<h2 id="peak_summary">Peak Summary</h2>


<strong>Switch between peaks below.</strong><br /><br />

<div>

  <!-- Nav tabs -->
  <ul class="nav nav-tabs" role="tablist">
  {% for peak in mapping_results %}
    <li role="presentation" peak="{{ peak.chrom }}:{{ peak.pos }}" class='peak-tab {% if loop.first %}active{% endif %}'><a href="#peak_{{ loop.index }}" aria-controls="peak_{{ loop.index }}" role="tab" data-toggle="tab">{{ peak.chrom }}:{{ peak.interval_start }}-{{ peak.interval_end }}</a></li>
    {% endfor %}
  </ul>

  <!-- Tab panes -->
  <div class="tab-content">
  {% for peak in mapping_results %}
    <div role="tabpanel" class="tab-pane fade {% if loop.first %}in active{% endif %}" id="peak_{{ loop.index }}">
    <br />
    <br />
    <ul>
      <li><strong>Peak Interval</strong> - {{ peak.chrom }}:{{ peak.interval_start }}-{{ peak.interval_end }}</li>
      <li><strong>Peak Position</strong> - {{ peak.chrom}}:{{ peak.pos }}</li>
    </ul>
    <div class="row">
    <div class="col-md-6">
      <table class="peak_summary table table-hover table-striped">
          <thead>
            <tr>
              <th>Gene Class</th>
              <th>All</th>
              <th>Moderate Effect</th>
              <th>High Effect</th>
            </tr>
          </thead>
          <tbody>
              <tr class="peak_{{ loop.index0 }}_Total">
                <td><strong>Total</strong></td>
                <td>-</td>
                <td>-</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_protein_coding">
                <td>Protein Coding</td>
                <td>-</td>
                <td>-</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_pseudogene">
                <td>pseudogene</td>
                <td>-</td>
                <td>-</td>
                <td>-</td>
              </tr>
              <tr>
              <td colspan = "4">
              * Predicted effects are only available for protein-coding genes.
              </td>
            </tbody>
          </table>
      </div>

      <div class="col-md-6">
        <table class="table table-hover table-striped">
        <thead>
        <tr>
        <th>Gene Class</th>
        <th>Count</th>
        </tr>
        </thead>
        <tbody>
              <tr class="peak_{{ loop.index0 }}_ncRNA">
                <td>ncRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_miRNA">
                <td>miRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_piRNA">
                <td>piRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_tRNA">
                <td>tRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_lincRNA">
                <td>lincRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_rRNA">
                <td>rRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_scRNA">
                <td>scRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_snoRNA">
                <td>snoRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_snRNA">
                <td>snRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_asRNA">
                <td>asRNA</td>
                <td>-</td>
              </tr>
          </tbody>
      </table>
      </div>
      </div>
    </div>
  {% endfor %}
  </div>

    <h3>Global Distribution</h3>

    <div id="map" class="col-md-8" style="height: 500px; width:100%"></div>
    <p>
      <img style="margin:10px; width:25px;" src="/static/img/icons/marker.png" /><strong>Reference (<span id="ref_gt"></span>)</strong>
      <img style="margin:10px; width:25px;" src="/static/img/icons/selected.png" /><strong>Alternative (<span id="alt_gt"></span>)</strong>
    </p>
    <h3>Tajima's D</h3>
      <div id="tajima" style="height: 500px;"></div>
          <div>

          </div>
   
    <p>Tajima's D across the most signficant QTL interval. Tajima's D tests whether a genomic region has the expected level of variation if it is only undergoing drift - values close to 0 indicate observed variation similar to expected levels, D&lt;0 suggests that rare alleles are present, which may have arose due to a recent selective sweep, population expansion after a recent bottleneck, or linkage to a swept gene, and D&gt;0 indicating balancing selection or sudden population contraction.</p>


  </div>

{% else %}

<div class="progress">
  <div class="progress-bar progress-bar-success progress-bar-striped active" role="progressbar"
  aria-valuenow="40" aria-valuemin="0" aria-valuemax="100" style="width:0%">
  </div>
</div>

{% endif %}

{% endblock %}


{% block script %}
<script>

{% if trait_data.status != "complete" %}
prog = Array()
prog["queue"] = ""
prog["Processing Phenotype Data"] = "10%"
prog["Performing Mapping"] = "20%"
prog["Processing Mapping"] = "60%"
prog["Plotting Figures"] = "75%"
prog["Processing Peaks"] = "80%"
prog["Transferring Data"] = "95%"
prog["complete"] = "100%"
prog["error"] = "100%"

var update_status = function(status) {
  if (status == "complete") {
    $(".progress").fadeOut();
    location.reload();
  } else if (status == "error") {
    $(".progress-bar").addClass("progress-bar-danger").html("error");
  } else if (status == "queue") {
    $(".progress-bar").css("width", "100%").addClass("progress-bar-info").html("Queued")
  } else {
    $(".progress-bar").removeClass("progress-bar-info").css("width", prog[status]).html(prog[status] + " (" + status + ")");
  }
}

function check_status() {
  $.ajax({
     url:"/report_progress/",
     data: JSON.stringify({
            "report_slug": "{{ report_slug }}",
            "trait_slug": "{{ trait_slug }}" }),
     method: "POST",
     contentType: 'application/json;charset=UTF-8'
     }
     ).done(function(msg) {
      msg = JSON.parse(msg);
      update_status(msg);
    });
}

$(document).ready(function() {
    $('#example').DataTable( {
        "ajax": "/cendr/static/content/data/interval_variants.tsv",
        "deferRender": true
    } );
} );

check_status();

window.setInterval(check_status, 3000);

{% else %}


// Set Gene Counts
$(document).ready(function() {
$.each($(".confidence_intervals > tbody > tr").children(":nth-child(4)"), function(k, v) {
  pos = $(v).siblings().first().children().html();
  chrom = pos.split(":")[0]
  start = pos.split(":")[1].split("-")[0].split(",").join("");
  end = pos.split(":")[1].split("-")[1].split(",").join("");
  gene_count = $(this);
  $.ajax({
  url: "/api/interval/" + chrom + "/" + start + "/" + end,
  contentType: 'application/json',
  success:function(msg) {
    console.log(msg);
    $(v).html(msg["total"]);
    $(v).next().html(msg["protein_coding"]);
    $(".peak_" + k + "_total").html("<strong>" + msg["total"] + "</strong>");
    biotypes = ["Total", "protein_coding", "pseudogene", "ncRNA", "miRNA", "piRNA", "asRNA", "snRNA", "snoRNA", "scRNA", "tRNA", "lincRNA", "rRNA"];
    $.each(biotypes, function(i, biotype) {
      // ALL
      $(".peak_" + k + "_" + biotype + " > td:nth-child(2)").html(msg["ALL_" + biotype]);
    });
    $.each(biotypes.slice(0,3), function(i, biotype) {
      // Moderate
      $(".peak_" + k + "_" + biotype + " > td:nth-child(3)").html(msg["MODERATE_" + biotype]);
      // High
      $(".peak_" + k + "_" + biotype + " > td:nth-child(4)").html(msg["HIGH_" + biotype]);
    })
  }
})
});
});

// Document ready
$('#report_nav').ready(function () {

    $('body').scrollspy({target: ".navbar", offset: 75}); 
    // Add smooth scrolling to all links inside a navbar
    $("#report_nav a").on('click', function(event){

      // Prevent default anchor click behavior
      event.preventDefault();

      // Store hash (#)
      var hash = this.hash;

      // Using jQuery's animate() method to add smooth page scroll
      // The optional number (800) specifies the number of milliseconds it takes to scroll to the specified area (the speed of the animation)
      $('html, body').animate({
        scrollTop: $(hash).offset().top - 74
      }, 800, function(){

        // Add hash (#) to URL when done scrolling (default click behavior)
        window.location.hash = hash;
      });
    });

});


//
// Genotype Location
//

var geo_gt = {{ geo_gt|safe }}

// Icons
var icon_norm = L.icon({
    iconUrl: '/static/img/icons/marker.png',
    iconSize: [24, 24],
    iconAnchor: [12, 16],
    popupAnchor: [-3, -76]
});

var icon_hover = L.icon({
    iconUrl: '/static/img/icons/selected.png',
    iconSize: [24, 24],
    iconAnchor: [12, 16],
    popupAnchor: [-3, -76]
});

gt_icon = {"0/0": icon_norm, "1/1": icon_hover}

ACCESS_TOKEN = 'pk.eyJ1IjoianByNyIsImEiOiJjaWhxbTRvdmswMDFzdGRtMDd0cGN6cG9lIn0.YzgwcnxBn6_DauMzz5fiQg';
MB_ATTR = 'Map data &copy; <a href="https://openstreetmap.org">OpenStreetMap</a> contributors, ' +
  '<a href="https://creativecommons.org/licenses/by-sa/2.0/">CC-BY-SA</a>, ' +
  'Imagery Â© <a href="http://mapbox.com">Mapbox</a>';
MB_URL = 'https://api.tiles.mapbox.com/v4/{id}/{z}/{x}/{y}.png?access_token=' + ACCESS_TOKEN;
OSM_URL = 'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png';
OSM_ATTRIB = '&copy; <a href="http://openstreetmap.org/copyright">OpenStreetMap</a> contributors';

var southWest = L.latLng(-90, -180),
  northEast = L.latLng(90, 180),
  bounds = L.latLngBounds(southWest, northEast);

var map = L.map('map', {minZoom:2, maxBounds: bounds}).setView([20, -100], 2);
L.tileLayer(MB_URL, {attribution: MB_ATTR, id: 'mapbox.streets', continuousWorld: false, worldCopyJump: true}).addTo(map);

var strain_layer = L.layerGroup();
var gt_set = {"0/0": "-", "1/1": "-"}



function update_geo_map(chrom, pos) {
  map.removeLayer(strain_layer);
  chrom_pos = chrom + ":" + pos;
  markers = [];
  $.each(geo_gt[chrom_pos], function(k, d) {
      m = L.marker([d.latitude, d.longitude], { icon: gt_icon[d.GT], tgt: d.TGT, gt: d.GT, title: d.isotype});
      markers.push(m);
      gt_set[d.GT] = d.TGT.split("/")[0];
  });
  strain_layer = L.layerGroup(markers).addTo(map);
  $("#ref_gt").html(gt_set["0/0"]);
  $("#alt_gt").html(gt_set["1/1"]);

}


//
// Tajima D
//

var data = [];
var layout = {
  margin: {
    t: 10,
    l: 40,
    r: 3
  },
  xaxis: {
    domain: [0, 1],
    title: "{{ mapping_results.0.chrom }}",
    range: [ {{ mapping_results.0.interval_start }}, {{ mapping_results.0.interval_end }} ],
    fixedrange: false,
    anchor: "x1"
  },
  yaxis: {
    range: [-4, 4],
    fixedrange: true
  },
  dragmode: "pan",
  showlegend: false,
  hovermode: 'closest'
};


function update_tajima_plot(chrom, start, end) {
  if (start < 0) {
    start = 0;
  }
  if (end < 0) {
    end = 0;
  }
  $.ajax({
    url: "/api/tajima/" + chrom + "/" + start + "/" + end,
    contentType: 'application/json',
    success:function(data) {
      data["mode"] = "markers";
      data["marker"] = {size: 5};
      layout.xaxis.title = chrom;
      layout.xaxis.range = [start, end];
      Plotly.newPlot('tajima', [data], layout, {displayModeBar: false, displaylogo: false});
  }
});
}


$("#tajima").bind('plotly_relayout', function(event, eventdata) {
  chrom = $(".peak-tab.active").attr("peak").split(":")[0];
  start = parseInt(eventdata["xaxis.range[0]"]);
  end = parseInt(eventdata["xaxis.range[1]"]);
  update_tajima_plot(chrom, start, end);
});


// Load initial map and Tajima
$(document).ready(function() {
  update_geo_map('{{ mapping_results.0.chrom }}' , {{ mapping_results.0.pos }});
  // Load initial position
  update_tajima_plot('{{ mapping_results.0.chrom }}', {{ mapping_results.0.interval_start }}, {{ mapping_results.0.interval_end }} )
});

// Monitor peak tabs
$('a[data-toggle="tab"]').on('shown.bs.tab', function (e) {
  peak = $(e.target).parent().attr("peak");
  chrom = peak.split(":")[0];
  pos = peak.split(":")[1];
  interval = $(".peak-tab.active").children("a").html()
  start = parseInt(interval.split(":")[1].split("-")[0]);
  end = parseInt(interval.split(":")[1].split("-")[1]);
  update_geo_map(chrom, pos);
  update_tajima_plot(chrom, start, end);
});

{% endif %}


</script>
{% endblock %}
