{% extends "_layouts/default.html" %}


{% block custom_head %}
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<script src="https://cdn.plot.ly/plotly-latest.min.js"></script>

<link rel="stylesheet" href="/static/css/leaflet.css" />
<script src="/static/js/leaflet.js"></script>
<script src="/static/js/leaflet-search.js"></script>



<script type="text/x-mathjax-config">
  MathJax.Hub.Config({tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}});
</script>
<script type="text/javascript" async
  src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_CHTML">
</script>


<style>

.report {
  width: 100%;
}

.height-report {
  height: 60%;
  width: 60%;
  display: block;
  margin-left: auto;
  margin-right: auto
}

#report_nav.affix {
    position: fixed;
    top: 0px;
    left: 0px;
    width: 100%;
}



/* Keep Header Scroll */
#gene_list table {
        width: 100%;
    }

#gene_list thead, #gene_list tbody, #gene_list tr, #gene_list td, #gene_list th { display: block; }

#gene_list tr:after {
    content: ' ';
    display: block;
    visibility: hidden;
    clear: both;
}

#gene_list thead th {
    height: 30px;

    /*text-align: left;*/
}

#gene_list tbody {
    height: 500px;
    overflow-y: auto;
}

#gene_list tbody td, #gene_list thead th {
    width: 14.28%;
    float: left;
}


</style>


{% endblock %}


{% block style %}
{% endblock %}

{% block content %}

<div class="row">
      <div class="col-md-12">

      <div class="btn-group" role="group">
        {% for trait in report_data %} 

          <a href="{{ url_for('trait_view', report_slug = report_url_slug, trait_slug = trait.trait_slug) }}" type="button" class="btnbtn-secondary btn-sm {% if trait.trait_slug == trait_slug %} btn-primary {% endif %}">{{ trait.trait_name }}</a>     
          
        {% endfor %}

        </div>
        <hr />
      </div>
</div>

{% if trait_data.status == "complete" %}
{% cache 50, report_slug, trait_slug %}
<div class="row">

<div class="col-md-12">

<div class="panel panel-default">
  <div class="panel-heading">
    <h3 class="panel-title">{{ trait_data["trait_name"] }}</h3>
  </div>
  <div class="panel-body">
  <div class="row">
  <div class='col-md-6'>
  <strong>Data</strong>
    <ul>
      {% if 'raw_phenotype.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/raw_phenotype.tsv">Raw Phenotype Data</a></strong></li>
      {% endif %}

      {% if 'processed_phenotype.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/processed_phenotype.tsv">Processed Phenotype Data</a></strong></li>
      {% endif %}

      {% if 'interval_variants.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/interval_variants.tsv">Interval Variants</a></strong></li>
      {% endif %}

      {% if 'raw_mapping.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/raw_mapping.tsv">Raw Mappings</a></strong></li>
      {% endif %}

      {% if 'processed_sig_mapping.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/processed_sig_mapping.tsv">Processed Significant Mappings</a></strong></li>
      {% endif %}

      {% if 'non_sig_mapping.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/non_sig_mapping.tsv">Nonsignificant Mappings</a></strong></li>
      {% endif %}
    </ul>
    </div>
    <div class='col-md-6'>
    <strong>Info</strong>
    <ul>
      <li><strong>Version</strong>: {{ trait_data.version }}</li>
      <li><strong>Availabity</strong>: {% if trait_data.release == 0 %}
                                            Public
                                         {% elif trait_data.release == 1 %}
                                            Embargo
                                         {% else %}
                                            Private
                                         {% endif %}</li>
    </ul>
    </div>
    </div>
  </div>
</div>

  </div>
</div>


<div id="report" data-toggle="tab">


<nav id="report_nav" data-offset-top="385" data-spy="affix" class="navbar navbar-default" style="z-index:10000">
  <div class="container-fluid">
   <!-- Collect the nav links, forms, and other content for toggling -->
    <div class="collapse navbar-collapse">
      <ul class="nav navbar-nav">
        <li><a href="#phenotype">Phenotype</a></li>
        <li><a href="#manhattan">Manhattan</a></li>
        {% if mapping_results|length > 0 %}
        <li><a href="#intervals">Intervals</a></li>
        <li><a href="#pxg">PxG</a></li>
        <li><a href="#ld">LD</a></li>
        <li><a href="#peak_summary">Peak Summary</a></li>
        {% endif %}
      </ul>
    </div><!-- /.navbar-collapse -->
  </div><!-- /.container-fluid -->
</nav>


<h2 id="phenotype" name="phenotype">Phenotype</h2>


<img class="report" src="{{ base_url }}/figures/phenotype_histogram-1.png" />
<strong>Histogram representation of your phenotype</strong>

{% if mapping_results|length > 0 %}

  Your phenotype is significantly associated with genetic variation present in the <em>C. elegans</em> population!

{% else %}

  <p>Unfortunately your phenotype was not significantly associated with genetic variation present in the <em>C. elegans</em> population. This could be due to noisy trait data - have you performed heritability analysis for your phenotype using our heritability strain panel? If you have and the heritability was found to be high for this trait, the trait might not have reached statistical significance because it is highly complex and more strains need to be phenotyped.</p>

  <p>If you have phenotyped the entire 152 wild-isolate collection, you can patiently wait for more isolates to be added to the collection, or generate an F2 recombinant inbred line (RIL) panel generated between strains with high and low phenotypes.</p>

{% endif %}


<a name="manhattan"></a>
<h2 id="manhattan">Manhattan Plot</h2>

{% if mapping_results|length > 0 %}

<img class="report" src="{{ base_url }}/figures/Manplot-1.png" />

A genome-wide representation of the association between variation in the <em>C. elegans</em> population and your phenotype. The x-axis corresponds to genomic position with chromosome number indicated as a roman numeral above each box. Dots represent single-nucleotide variants (SNV) present in the <em>C. elegans</em> population. The y-axis corresponds to the level of significance for the association test. Blue dots represent SNVs that are above the defined significance threshold, which is the thick gray line. Red boxes surrounding blue dots represent the QTL genomic region of interest, which we define as plus/minus 50 SNVs from the last signifincant SNV. The default threshold for significance is the Bonferroni-corrected value $(-log_{10}\frac{0.05}{\#SNVs})$ and is usually around 5.5.

{% else %}

<img class="report" src="{{ base_url }}/figures/non-sig Manhattan Plot-1.png" />

A genome-wide representation of the association between variation in the <em>C. elegans</em> population and your phenotype. The x-axis corresponds to genomic position with chromosome number indicated as a roman numeral above each box. Dots represent single-nucleotide variants (SNV) present in the <em>C. elegans</em> population. The y-axis corresponds to the level of significance for the association test. The red line corresponds to the Bonferroni-corrected significance threshold, which is defined as $(-log_{10}\frac{0.05}{\#SNVs})$.

{% endif %}


{% if mapping_results|length > 0 %}
<h3 id="intervals">QTL Genomic Region of Interest</h3>

<table class="table table-hover table-striped confidence_intervals">
<thead>
  <tr>
    <th>Interval</th>
    <th>Peak Position</th>
    <th>-log<sub>10</sub>p</th>
    <th>Total</th>
    <th>Protein Coding</th>
    <th>Moderate Impact</th>
    <th>High Impact</th>
  </tr>
</thead>
<tbody>
{% for row in mapping_results %}
<tr chrom="{{ row.chrom }}" start="{{ row.interval_start }}" end="{{ row.interval_end }}">
  <td><a href="http://www.wormbase.org/tools/genome/gbrowse/c_elegans_PRJNA13758/?name={{ row.chrom}}:{{ row.interval_start }}..{{ row.interval_end }}" target="_blank">{{ row.chrom }}:{{ "{:,.0f}".format(row.interval_start) }}-{{ "{:,.0f}".format(row.interval_end) }}</a></td>
  <td><a href="http://www.wormbase.org/tools/genome/gbrowse/c_elegans_PRJNA13758/?name={{ row.chrom}}:{{ row.interval_start }}..{{ row.interval_end }}" target="_blank">{{row.chrom}}:{{ "{:,.0f}".format(row.pos) }}</a></td>
  <td>{{ row.log10p|round(3) }}</td>
  <td class="peaksum_{{ loop.index0 }}_Total">-</td>
  <td class="peaksum_{{ loop.index0 }}_protein_coding">-</td>
  <td class="peaksum_{{ loop.index0 }}_MODERATE">-</td>
  <td class="peaksum_{{ loop.index0 }}_HIGH">-</td>
</tr>
{% endfor %}
</tbody>
</table>
<small>Moderate and High impact variants are terms used to describe the predicted severity of a variant as assigned by SnpEff. More information regarding the types of variants classified as Moderate or High impact is <a href="{{ url_for('help_item', filename = 'Variant-Prediction') }}">available in the help section</a>.</small>


<h3 id="pxg">Phenotype-by-Genotype</h3>
<br />
<img class="report" src="{{ base_url }}/figures/PxGplot-1.png" />

<p>Phenotypic distributions represented as box plots are split by the genotype at the most significant SNV for a given QTL. The phenotype is on the y-axis, and the genotype is on the x-axis. The chromosome and position for the plotted SNV are indicated as roman numerals above each panel. REF refers to the reference N2 genotype, and ALT refers to the alternative variant phenotype. All of the SNVs in our data set are biallelic, so only two classes for any given SNV site exist.</p>

<h3 id="ld">Linkage Disequilibrium</h3>

<img class="height-report" src="{{ base_url }}/figures/LDplot-1.png" />

<p>The Linkage Disequilibrium (LD) measure plotted is the correlation between peak markers $r = \frac{-D}{\sqrt{p(A) * p(a) * p(B) * p*(b)}}$, where D raw difference in frequency between the observed number of AB pairs and the expected number and A, B, a, and b refer to alleles at the two loci. An LD value of 0.8 or higher suggests that the two peak markers are not segregating randomly.</p>
<br />
<h2 id="peak_summary">Peak Summary</h2>

<strong>Switch between peaks below.</strong><br /><br />

<div>

  <!-- Nav tabs -->
  <ul class="nav nav-tabs" role="tablist">
  {% for peak in mapping_results %}
    <li role="presentation" peak="{{ peak.chrom }}:{{ peak.pos }}" class='peak-tab {% if loop.first %}active{% endif %}'><a href="#peak_{{ loop.index }}" aria-controls="peak_{{ loop.index }}" role="tab" data-toggle="tab">{{ peak.chrom }}:{{ peak.interval_start }}-{{ peak.interval_end }}</a></li>
    {% endfor %}
  </ul>

  <!-- Tab panes -->
  <div class="tab-content">
  {% for peak in mapping_results %}
    <div role="tabpanel" class="tab-pane fade {% if loop.first %}in active{% endif %}" id="peak_{{ loop.index }}">
    <br />
    <br />
    <ul>
      <li><strong>Peak Interval</strong> - {{ peak.chrom }}:{{ peak.interval_start }}-{{ peak.interval_end }}</li>
      <li><strong>Peak Position</strong> - {{ peak.chrom}}:{{ peak.pos }}</li>
    </ul>
    <div class="row">
    <div class="col-md-6">
      <table class="peak_summary table table-hover table-striped">
          <thead>
            <tr>
              <th>Gene Class</th>
              <th># Genes</th>
              <th>>=1 Moderate Impact</th>
              <th>>=1 High Impact</th>
            </tr>
          </thead>
          <tbody>
              <tr class="peak_{{ loop.index0 }}_Total">
                <td><strong>Total</strong></td>
                <td>-</td>
                <td>-</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_protein_coding">
                <td>Protein Coding</td>
                <td>-</td>
                <td>-</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_pseudogene">
                <td>pseudogene</td>
                <td>-</td>
                <td>-</td>
                <td>-</td>
              </tr>
              <tr>
              <td colspan = "4">
              <p style="text-align">Numbers listed under Moderate and High Impact correspond to the number of genes within the interval that have at least 1 variant classified as having a moderate or high impact.</p>
              <p style="text-align: justify">Moderate and High impact variants are terms used to describe the predicted severity of a variant as assigned by SnpEff. More information regarding the types of variants classified as Moderate or High impact is <a href="{{ url_for('help_item', filename = 'Variant-Prediction') }}">available in the help section</a>.</p>
              </td>
            </tbody>
          </table>
      </div>

      <div class="col-md-6">
        <table class="table table-hover table-striped">
        <thead>
        <tr>
        <th>Gene Class</th>
        <th># Genes</th>
        </tr>
        </thead>
        <tbody>
              <tr class="peak_{{ loop.index0 }}_ncRNA">
                <td>ncRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_miRNA">
                <td>miRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_piRNA">
                <td>piRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_tRNA">
                <td>tRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_lincRNA">
                <td>lincRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_rRNA">
                <td>rRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_scRNA">
                <td>scRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_snoRNA">
                <td>snoRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_snRNA">
                <td>snRNA</td>
                <td>-</td>
              </tr>
              <tr class="peak_{{ loop.index0 }}_asRNA">
                <td>asRNA</td>
                <td>-</td>
              </tr>
          </tbody>
      </table>
      </div>
      </div>
    </div>
  {% endfor %}
  </div>

    <h3>Genes</h3>

<table class="table table-striped table-hover" id="gene_list">
  <thead>
    <tr>
      <th>CHROM</th>
      <th>start</th>
      <th>end</th>
      <th>locus</th>
      <th>Sequence</th>
      <th>Wormbase ID</th>
      <th>Biotype</th>
    </tr>
  </thead>
  <tbody>
  </tbody>
</table>


    <h3>Global Distribution</h3>

    <div id="map" class="col-md-8" style="height: 500px; width:100%"></div>
    <p>
      <img style="margin:10px; width:25px;" src="/static/img/icons/marker.png" /><strong>Reference (<span id="ref_gt"></span>)</strong>
      <img style="margin:10px; width:25px;" src="/static/img/icons/selected.png" /><strong>Alternative (<span id="alt_gt"></span>)</strong>
    </p>

    <div class="progress">
      <div class="progress-bar progress-bar-warning" id="var_percent" role="progressbar" style="width:10%; background-color:#0066FF">
        Variant
        <p class="text"></p>
      </div>
      <div class="progress-bar progress-bar-success" id= "ref_percent" role="progressbar" style="width:80%; background-color:#C0382A">
        Reference
        <p class="text"></p>
      </div>
    </div>

    <h3>Tajima's D</h3>
      <div id="tajima" style="height: 500px;"></div>
          <div>

          </div>
   
    <p>Tajima's D is plotted across the most significant QTL genomic region of interest. Tajima's D suggests if a genomic region has an excess or paucity of rare variation. Values close to zero indicate that variation is drifting and not selected. Values less than zero suggest that rare variants are present, which could have arose during a selective sweep, population expansion after a bottleneck, or linkage to a swept allele. Values greater than zero suggest that balancing selection could have occurred or recent population contractions.</p>


  </div>

{% endif %}
{% endcache %}
{% else %}

<div class="progress">
  <div class="progress-bar progress-bar-success progress-bar-striped active" role="progressbar"
  aria-valuenow="40" aria-valuemin="0" aria-valuemax="100" style="width:0%">
  </div>
</div>
{% endif %}

{% endblock %}


{% block script %}
<script>

{% if trait_data.status != "complete" %}
prog = Array()
prog["queue"] = ""
prog["Processing Phenotype Data"] = "10%"
prog["Performing Mapping"] = "20%"
prog["Processing Mapping"] = "60%"
prog["Plotting Figures"] = "75%"
prog["Processing Peaks"] = "80%"
prog["Transferring Data"] = "95%"
prog["complete"] = "100%"
prog["error"] = "100%"

var update_status = function(status) {
  if (status == "complete") {
    $(".progress").fadeOut();
    location.reload();
  } else if (status == "error") {
    $(".progress-bar").addClass("progress-bar-danger").html("error");
  } else if (status == "queue") {
    $(".progress-bar").css("width", "100%").addClass("progress-bar-info").html("Queued")
  } else {
    $(".progress-bar").removeClass("progress-bar-info").css("width", prog[status]).html(prog[status] + " (" + status + ")");
  }
}

function check_status() {
  $.ajax({
     url:"/report_progress/",
     data: JSON.stringify({
            "report_slug": "{{ report_slug }}",
            "trait_slug": "{{ trait_slug }}" }),
     method: "POST",
     contentType: 'application/json;charset=UTF-8'
     }
     ).done(function(msg) {
      msg = JSON.parse(msg);
      update_status(msg);
    });
}

check_status();

window.setInterval(check_status, 3000);

{% else %}

{% if mapping_results|length > 0 %}
// Set Gene Counts
$(document).ready(function() {
$.each($(".confidence_intervals > tbody > tr").children(":nth-child(4)"), function(k, v) {
  pos = $(v).siblings().first().children().html();
  chrom = pos.split(":")[0]
  start = pos.split(":")[1].split("-")[0].split(",").join("");
  end = pos.split(":")[1].split("-")[1].split(",").join("");
  gene_count = $(this);
  $.ajax({
  url: "/api/interval/" + chrom + "/" + start + "/" + end,
  contentType: 'application/json',
  success:function(msg) {
    $(v).html(msg["total"]);
    $(v).next().html(msg["protein_coding"]);
    $(".peak_" + k + "_total").html("<strong>" + msg["total"] + "</strong>");
    biotypes = ["Total", "protein_coding", "pseudogene", "ncRNA", "miRNA", "piRNA", "asRNA", "snRNA", "snoRNA", "scRNA", "tRNA", "lincRNA", "rRNA"];

    $(".peaksum_" + k + "_Total").html(msg["ALL_Total"]);
    $(".peaksum_" + k + "_protein_coding").html(msg["ALL_protein_coding"]);
    $(".peaksum_" + k + "_MODERATE").html(msg["MODERATE_Total"]);
    $(".peaksum_" + k + "_HIGH").html(msg["HIGH_Total"]);

    $.each(biotypes, function(i, biotype) {
      // ALL
      $(".peak_" + k + "_" + biotype + " > td:nth-child(2)").html(msg["ALL_" + biotype]);
    });
    $.each(biotypes.slice(0,3), function(i, biotype) {
      // Moderate
      $(".peak_" + k + "_" + biotype + " > td:nth-child(3)").html(msg["MODERATE_" + biotype]);
      // High
      $(".peak_" + k + "_" + biotype + " > td:nth-child(4)").html(msg["HIGH_" + biotype]);
    })
  }
})
});
});

// Document ready
$('#report_nav').ready(function () {

    $('body').scrollspy({target: ".navbar", offset: 75}); 
    // Add smooth scrolling to all links inside a navbar
    $("#report_nav a").on('click', function(event){

      // Prevent default anchor click behavior
      event.preventDefault();

      // Store hash (#)
      var hash = this.hash;

      // Using jQuery's animate() method to add smooth page scroll
      // The optional number (800) specifies the number of milliseconds it takes to scroll to the specified area (the speed of the animation)
      $('html, body').animate({
        scrollTop: $(hash).offset().top - 74
      }, 800, function(){

        // Add hash (#) to URL when done scrolling (default click behavior)
        window.location.hash = hash;
      });
    });

});


//
// Genotype Location
//

var geo_gt = {{ geo_gt|safe }}

// Icons
var icon_norm = L.icon({
    iconUrl: '/static/img/icons/marker.png',
    iconSize: [24, 24],
    iconAnchor: [12, 16],
    popupAnchor: [-3, -76]
});

var icon_hover = L.icon({
    iconUrl: '/static/img/icons/selected.png',
    iconSize: [24, 24],
    iconAnchor: [12, 16],
    popupAnchor: [-3, -76]
});

gt_icon = {"0/0": icon_norm, "1/1": icon_hover}

ACCESS_TOKEN = 'pk.eyJ1IjoianByNyIsImEiOiJjaWhxbTRvdmswMDFzdGRtMDd0cGN6cG9lIn0.YzgwcnxBn6_DauMzz5fiQg';
MB_ATTR = 'Map data &copy; <a href="https://openstreetmap.org">OpenStreetMap</a> contributors, ' +
  '<a href="https://creativecommons.org/licenses/by-sa/2.0/">CC-BY-SA</a>, ' +
  'Imagery Â© <a href="http://mapbox.com">Mapbox</a>';
MB_URL = 'https://api.tiles.mapbox.com/v4/{id}/{z}/{x}/{y}.png?access_token=' + ACCESS_TOKEN;
OSM_URL = 'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png';
OSM_ATTRIB = '&copy; <a href="http://openstreetmap.org/copyright">OpenStreetMap</a> contributors';

var southWest = L.latLng(-90, -180),
  northEast = L.latLng(90, 180),
  bounds = L.latLngBounds(southWest, northEast);

var map = L.map('map', {minZoom:2, maxBounds: bounds}).setView([20, -100], 2);
L.tileLayer(MB_URL, {attribution: MB_ATTR, id: 'mapbox.streets', continuousWorld: false, worldCopyJump: true}).addTo(map);

var strain_layer = L.layerGroup();
var gt_set = {"0/0": "-", "1/1": "-"}



function update_geo_map(chrom, pos) {
  map.removeLayer(strain_layer);
  chrom_pos = chrom + ":" + pos;
  markers = [];
  ref_counter = 0;
  alt_counter = 0;
  other = 0
  $.each(geo_gt[chrom_pos], function(k, d) {

      if (d.GT == '0/0') {
        ref_counter += 1;
      } else if (d.GT == '1/1') {
        alt_counter += 1;
      }
      else{
        other += 1
      }
      total = ref_counter + alt_counter + other
      refer_percent = (ref_counter/total) * 100
      alter_percent = (alt_counter/total) * 100

      m = L.marker([d.latitude, d.longitude], { icon: gt_icon[d.GT], tgt: d.TGT, gt: d.GT, title: d.isotype});
      markers.push(m);
      gt_set[d.GT] = d.TGT.split("/")[0];
  });
  strain_layer = L.layerGroup(markers).addTo(map);
  $("#ref_gt").html(gt_set["0/0"]);
  $("#alt_gt").html(gt_set["1/1"]);

  $("#ref_percent").width(refer_percent+'%');
  $("#var_percent").width(alter_percent+'%');

}


//
// Tajima D
//

var data = [];
var layout = {
  margin: {
    t: 10,
    l: 40,
    r: 3
  },
  xaxis: {
    domain: [0, 1],
    title: "{{ mapping_results.0.chrom }}",
    range: [ {{ mapping_results.0.interval_start }}, {{ mapping_results.0.interval_end }} ],
    fixedrange: false,
    anchor: "x1"
  },
  yaxis: {
    range: [-4, 4],
    fixedrange: true
  },
  dragmode: "pan",
  showlegend: false,
  hovermode: 'closest'
};


function update_tajima_plot(chrom, start, end) {
  start = start - 500000;
  end = end + 500000;
  if (start < 0) {
    start = 0;
  }
  if (end < 0) {
    end = 0;
  }
  $.ajax({
    url: "/api/variant/tajima/" + chrom + "/" + start + "/" + end,
    contentType: 'application/json',
    success:function(data) {
      data["mode"] = "markers";
      data["marker"] = {size: 5};
      layout.xaxis.title = chrom;
      layout.xaxis.range = [start, end];
      Plotly.newPlot('tajima', [data], layout, {displayModeBar: false, displaylogo: false});
  }
});
}


$("#tajima").bind('plotly_relayout', function(event, eventdata) {
  chrom = $(".peak-tab.active").attr("peak").split(":")[0];
  start = parseInt(eventdata["xaxis.range[0]"]);
  end = parseInt(eventdata["xaxis.range[1]"]);
  update_tajima_plot(chrom, start, end);
});


// Gene List
function update_gene_list(chrom, start, end) {
  $("#gene_list > tbody").html("");
    $.ajax({
    url: "/api/genelist/" + chrom + "/" + start + "/" + end,
    contentType: 'application/json',
    success:function(msg) {
      $.each(msg["gene_list"], function(k,v) {

      if (v["locus"] != "") {
        v["locus"] = "<a href='http://www.wormbase.org/species/c_elegans/gene/" + v["locus"] + "'>" + v["locus"] + "</a>";
      }
      if (v["Name"] != "") {
        v["Name"] = "<a href='http://www.wormbase.org/species/c_elegans/gene/" + v["Name"] + "'>" + v["Name"] + "</a>";
      }


      items = [v["CHROM"],
               Number(v["start"]).toLocaleString('en'),
               Number(v["end"]).toLocaleString('en'),
               v["locus"],
               v["sequence_name"],
               v["Name"],
               v["biotype"]];
      $("#gene_list > tbody").append("<tr><td>" + items.join("</td><td>") + "</td></tr>");
      });

    // Enable links
    $( "#gene_list a" ).on( "click", function() {
        window.location = $(this).attr('href');
    });


    }
  });

}



// Load initial map and Tajima
$(document).ready(function() {
  update_geo_map('{{ mapping_results.0.chrom }}' , {{ mapping_results.0.pos }});
  // Load initial position
  update_tajima_plot('{{ mapping_results.0.chrom }}', {{ mapping_results.0.interval_start }}, {{ mapping_results.0.interval_end }} );
  update_gene_list('{{ mapping_results.0.chrom }}', {{ mapping_results.0.interval_start }}, {{ mapping_results.0.interval_end }});
});

// Monitor peak tabs
$('a[data-toggle="tab"]').on('shown.bs.tab', function (e) {
  peak = $(e.target).parent().attr("peak");
  chrom = peak.split(":")[0];
  pos = peak.split(":")[1];
  interval = $(".peak-tab.active").children("a").html()
  start = parseInt(interval.split(":")[1].split("-")[0]);
  end = parseInt(interval.split(":")[1].split("-")[1]);
  update_geo_map(chrom, pos);
  update_tajima_plot(chrom, start, end);
  update_gene_list(chrom, start, end);
});

{% endif %}
{% endif %}

</script>
{% endblock %}
