{% extends "_layouts/default.html" %}


{% block custom_head %}
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta name="viewport" content="width=device-width, initial-scale=1" />

<script type="text/javascript" src="https://www.gstatic.com/charts/loader.js"></script>
<link rel="stylesheet" href="/static/css/leaflet.css" />
<script src="/static/js/leaflet.js"></script>
<script src="/static/js/leaflet-search.js"></script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}});
</script>
<script type="text/javascript" async
  src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_CHTML">
</script>


<style>

.report {
  width: 100%;
}

.height-report {
  height: 60%;
  width: 60%;
  display: block;
  margin-left: auto;
  margin-right: auto
}

#report_nav.affix {
    position: fixed;
    top: 0px;
    left: 0px;
    width: 100%;
}



/* Keep Header Scroll */
#variant_list table {
    width: 100%;
}

#variant_list thead, #variant_list tbody, #variant_list tr, #variant_list td, #variant_list th { display: block; }

#variant_list tr:after {
    content: ' ';
    display: block;
    visibility: hidden;
    clear: both;
}

#gene_list thead th, #variant_list thead th {
    height: 30px;
    /*text-align: left;*/
}

#gene_list tbody, #variant_list tbody {
    height: 500px;
    overflow-y: auto;
}

#gene_list tbody td, #gene_list thead th, #variant_list tbody td, #variant_list thead th {
    width: 16.66%;
    float: left;
}


</style>


{% endblock %}


{% block style %}
{% endblock %}

{% block content %}

<div class="row">
      <div class="col-md-12">

      <div class="btn-group" role="group">
        {% for trait in report_data %} 

          <a href="{{ url_for('trait_view', report_slug = report_url_slug, trait_slug = trait.trait_slug) }}" type="button" class="btn btn-sm {% if trait.trait_slug == trait_slug %} btn-primary {% else %} btn-link {% endif %}">{% if trait.interval_start %}<strong>{{ trait.trait_name }}</strong>{% else %}{{ trait.trait_name }}{% endif %}</a>     
          
        {% endfor %}

      </div>
      </div>
</div>

{% if trait_data.status == "complete" %}
{% cache 50, report_slug, trait_slug %}
<div class="row">

<div class="col-md-12">
<br /><br />
<div class="panel panel-default">
  <div class="panel-body">
  <div class="row">
  <div class='col-md-6'>
  <strong>Data</strong>
    <ul>
      {% if 'phenotype.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/phenotype.tsv">Phenotype Data</a></strong></li>
      {% endif %}

      {% if 'mapping.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/mapping.tsv">Mapping</a></strong></li>
      {% endif %}

      {% if mapping_results|length > 0 %}
      <li><strong><a href='{{ url_for('interval_download', report_slug = report_slug, trait_slug = trait_slug) }}'>Variant Correlation</a></strong></li>
      {% endif %}

    </ul>
    </div>
    <div class='col-md-6'>
    <strong>Info</strong>
    <ul>
      <li><strong>Version</strong>: {{ report_data[0].version }}</li>
      <li><strong>Availabity</strong>: {% if report_data[0].release == 0 %}
                                            Public
                                         {% elif report_data[0].release == 1 %}
                                            Embargo
                                         {% else %}
                                            Private
                                         {% endif %}</li>
    </ul>
    </div>
    </div>
  </div>
</div>

  </div>
</div>

<div id="report" data-toggle="tab">


<nav id="report_nav" data-offset-top="385" data-spy="affix" class="navbar navbar-default" style="z-index:10000">
  <div class="container-fluid">
   <!-- Collect the nav links, forms, and other content for toggling -->
    <div class="collapse navbar-collapse">
      <ul class="nav navbar-nav">
        <li><a href="#phenotype">Phenotype</a></li>
        <li><a href="#manhattan">Manhattan</a></li>
        {% if mapping_results|length > 0 %}
        <li><a href="#intervals">Intervals</a></li>
        <li><a href="#pxg">PxG</a></li>
        {% if mapping_results|length > 1 %}
        <li><a href="#ld">LD</a></li>
        {% endif %}
        <li><a href="#peak_summary">Peak Summary</a></li>
        {% endif %}
      </ul>
    </div><!-- /.navbar-collapse -->
  </div><!-- /.container-fluid -->
</nav>


<h2 id="phenotype" name="phenotype">Phenotype</h2>

<img class="report" src="{{ base_url }}/figures/Phenotype.png" />
<strong>Histogram representation of your phenotype</strong>

{% if mapping_results|length > 0 %}

  Your phenotype is significantly associated with genetic variation present in the <em>C. elegans</em> population!

{% else %}

  <p>Unfortunately your phenotype was not significantly associated with genetic variation present in the <em>C. elegans</em> population. This could be due to noisy trait data - have you performed heritability analysis for your phenotype using our heritability strain panel? If you have and the heritability was found to be high for this trait, the trait might not have reached statistical significance because it is highly complex and more strains need to be phenotyped.</p>

  <p>If you have phenotyped the entire 152 wild-isolate collection, you can patiently wait for more isolates to be added to the collection, or generate an F2 recombinant inbred line (RIL) panel generated between strains with high and low phenotypes.</p>

{% endif %}


<a name="manhattan"></a>
<h2 id="manhattan">Manhattan Plot</h2>

{% if mapping_results|length > 0 %}

<img class="report" src="{{ base_url }}/figures/Manhattan.png" />

A genome-wide representation of the association between variation in the <em>C. elegans</em> population and your phenotype. The x-axis corresponds to genomic position with chromosome number indicated as a roman numeral above each box. Dots represent single-nucleotide variants (SNV) present in the <em>C. elegans</em> population. The y-axis corresponds to the level of significance for the association test. Blue dots represent SNVs that are above the defined significance threshold, which is the thick gray line. Red boxes surrounding blue dots represent the QTL genomic region of interest, which we define as plus/minus 50 SNVs from the last signifincant SNV. The default threshold for significance is the Bonferroni-corrected value $(-log_{10}\frac{0.05}{\#SNVs})$ and is usually around 5.5.

{% else %}

<img class="report" src="{{ base_url }}/figures/Manhattan.png" />

A genome-wide representation of the association between variation in the <em>C. elegans</em> population and your phenotype. The x-axis corresponds to genomic position with chromosome number indicated as a roman numeral above each box. Dots represent single-nucleotide variants (SNV) present in the <em>C. elegans</em> population. The y-axis corresponds to the level of significance for the association test. The red line corresponds to the Bonferroni-corrected significance threshold, which is defined as $(-log_{10}\frac{0.05}{\#SNVs})$.

{% endif %}

{% if mapping_results|length > 0 %}
<h3 id="intervals">QTL Genomic Region of Interest</h3>

<table class="table table-hover table-striped confidence_intervals" >
<col width="15">
<col width="10">
<col width="5">
<col width="14">
<col width="14">
<col width="14">
<col width="14">
<col width="14">
<thead>
  <tr>
    <th>Interval</th>
    <th class='text-right'>Peak Position</th>
    <th class='text-right'>-log<sub>10</sub>p</th>
    <th class='text-right'><abbr data-toggle="tooltip"  data-placement="bottom" title="Variance Explained">V.E.</abbr></abbr></th>
    <th class='text-right'>Variants</th>
    <th class='text-right'><abbr data-toggle="tooltip"  data-placement="bottom" title="Number of coding genes with at least 1 variant">Coding</abbr></th>
    <th class='text-right'><abbr data-toggle="tooltip"  data-placement="bottom" title="Number of genes with >= 1 moderate impact variant">Moderate</abbr></th>
    <th class='text-right'><abbr data-toggle="tooltip"  data-placement="bottom" title="Number of genes with >= 1 high impact variant">High</abbr></th>
  </tr>
</thead>
<tbody>
{% for row in mapping_results %}
{% set peak_index = loop.index0 %}
<tr chrom="{{ row.chrom }}" start="{{ row.interval_start }}" end="{{ row.interval_end }}" class="text-right">
  <td class="text-left"><a href="http://www.wormbase.org/tools/genome/gbrowse/c_elegans_PRJNA13758/?name={{ row.chrom}}:{{ row.interval_start }}..{{ row.interval_end }}" target="_blank">{{ row.chrom }}:{{ "{:,.0f}".format(row.interval_start) }}-{{ "{:,.0f}".format(row.interval_end) }}</a></td>
  <td class='text-right'><a href="http://www.wormbase.org/tools/genome/gbrowse/c_elegans_PRJNA13758/?name={{ row.chrom}}:{{ row.interval_start }}..{{ row.interval_end }}" target="_blank">{{row.chrom}}:{{ "{:,.0f}".format(row.pos) }}</a></td>
  <td>{{ row.log10p|round(2) }}</td>
  <td>{{ row.variance_explained|round(3) * 100 }}%</td>
  <td data='msg.variants' class='dat_{{ peak_index }}'>-</td>
  <td data='msg.gene_w_variants_biotype.protein_coding' class='dat_{{ peak_index }}'>-</td>
  <td data='msg.gene_by_impact.total.MODERATE' class='dat_{{ peak_index }}'>-</td>
  <td data='msg.gene_by_impact.total.HIGH' class='dat_{{ peak_index }}'>-</td>
</tr>
{% endfor %}
</tbody>
</table>
<small>Moderate and High impact variants are terms used to describe the predicted severity of a variant as assigned by SnpEff. More information regarding the types of variants classified as Moderate or High impact is <a href="{{ url_for('help_item', filename = 'Variant-Prediction') }}">available in the help section</a>.</small>


<h3 id="pxg">Phenotype-by-Genotype</h3>
<br />
<img class="report" src="{{ base_url }}/figures/PxG.png" />

<p>Phenotypic distributions represented as box plots are split by the genotype at the most significant SNV for a given QTL. The phenotype is on the y-axis, and the genotype is on the x-axis. The chromosome and position for the plotted SNV are indicated as roman numerals above each panel. REF refers to the reference N2 genotype, and ALT refers to the alternative variant phenotype. All of the SNVs in our data set are biallelic, so only two classes for any given SNV site exist.</p>

{% if mapping_results|length > 1 %}

<h3 id="ld">Linkage Disequilibrium</h3>

<img class="height-report" src="{{ base_url }}/figures/LD.png" />

<p>The Linkage Disequilibrium (LD) measure plotted is the correlation between peak markers $r = \frac{-D}{\sqrt{p(A) * p(a) * p(B) * p*(b)}}$, where D raw difference in frequency between the observed number of AB pairs and the expected number and A, B, a, and b refer to alleles at the two loci. An LD value of 0.8 or higher suggests that the two peak markers are not segregating randomly.</p>
<br />
{% endif %}
<h2 id="peak_summary">Peak Summary</h2>

<strong>Switch between peaks below.</strong><br /><br />

<div>

  <!-- Nav tabs -->
  <ul class="nav nav-tabs" role="tablist">
  {% for peak in mapping_results %}
    <li role="presentation" peak="{{ peak.chrom }}:{{ peak.pos }}" class='peak-tab {% if loop.first %}active{% endif %}'><a href="#peak_{{ loop.index }}" aria-controls="peak_{{ loop.index }}" role="tab" data-toggle="tab">{{ peak.chrom }}:{{ peak.interval_start }}-{{ peak.interval_end }}</a></li>
    {% endfor %}
  </ul>

  <!-- Tab panes -->
  <div class="tab-content">
  {% for peak in mapping_results %}
  {% set peak_index = loop.index0 %}
    <div role="tabpanel" class="tab-pane fade {% if loop.first %}in active{% endif %}" id="peak_{{ loop.index }}">
    <br />
    <br />
    <ul>
      <li><strong>Peak Interval</strong> &nbsp;{{ peak.chrom }}:{{ peak.interval_start }}-{{ peak.interval_end }}</li>
      <li><strong>Peak Position</strong> &nbsp;{{ peak.chrom}}:{{ peak.pos }}</li>
      <li><strong>Variance Explained</strong> &nbsp;{{ peak.variance_explained|round(3) * 100 }}%</li>
    </ul>
    <div class="row">
    <div class="col-md-12">
      <table class="peak_summary table table-hover table-striped" style="table-layout: fixed">
          <thead>
            <tr>
              <th>Biotype</th>
              <th># Genes</th>
              <th># Genes w/ variants</th>
              <th>>=1 Moderate Impact</th>
              <th>>=1 High Impact</th>
              <th># Variants</th>
            </tr>
          </thead>
          <tbody class='peak_{{ loop.index0 }} text-right'>
              <tr>
                <td class='text-left'><strong>Total</strong></td>
                <td data='msg.gene_biotype.total' class='dat_{{ peak_index }}'>-</td>
                <td data='msg.gene_w_variants_biotype.total' class='dat_{{ peak_index }}'>-</td>
                <td data='msg.gene_by_impact.total.MODERATE' class='dat_{{ peak_index }}'>-</td>
                <td data='msg.gene_by_impact.total.HIGH' class='dat_{{ peak_index }}'>-</td>
                <td data='msg.variants' class='dat_{{ peak_index }}'>-</td>
              </tr>
              {% for i in ["protein_coding", "pseudogene", "ncRNA", "miRNA", "piRNA", "tRNA", "lincRNA", "rRNA", "scRNA", "snoRNA", "snRNA", "asRNA", "Coding", "Noncoding"] %}
              <tr>
                <td class='text-left'>
                {% if i in biotypes.keys() %}
                  <abbr data-toggle='tooltip' data-placement='top' data-original-title='{{ biotypes[i] }}'>{{ i }}</abbr>
                {% else %}
                {{ i }}
                {% endif %}
                </td>
                <td data='msg.gene_biotype.biotype_summary.{{ i }}' class='dat_{{ peak_index }}'>-</td>
                <td data='msg.gene_w_variants_biotype.{{ i }}' class='dat_{{ peak_index }}'>-</td>
                <td data='msg.gene_by_impact.MODERATE.{{ i }}' class='dat_{{ peak_index }}'>-</td>
                <td data='msg.gene_by_impact.HIGH.{{ i }}' class='dat_{{ peak_index }}'>-</td>
                <td data='msg.variant_in_biotype.{{ i }}' class='dat_{{ peak_index }}'>-</td>
              </tr>
              {% endfor %}
              <tr>
                <td colspan = "6">
                <p style="text-align">Numbers listed under Moderate and High Impact correspond to the number of genes within the interval that have at least 1 variant classified as having a moderate or high impact.</p>
                <p style="text-align: justify">Moderate and High impact variants are terms used to describe the predicted severity of a variant as assigned by SnpEff. More information regarding the types of variants classified as Moderate or High impact is <a href="{{ url_for('help_item', filename = 'Variant-Prediction') }}">available in the help section</a>.</p>
                </td>
              </tr>
        </tbody>
      </table>
      </div>
    </div>
      

<h3>Variants</h3>


<table class="table table-hover table-striped" id="variant_list" style="table-layout: fixed">
    <thead>
        <tr>
            <th>Gene Name</th>
            <th>Wormbase ID</th>
            <th>Biotype</th>
            <th><abbr data-toggle='tooltip' data-placement='top' data-original-title='The total number of variants with Moderate/High predicted effects'>Total Variants</abbr></th>
            <th><abbr data-toggle='tooltip' data-placement='top' data-original-title='The Maximum correlation among all variants examined'>Max. Correlation</abbr></th>
            <th>Toggle Variants</th>
        </tr>
    </thead>
<tbody>
{% for gene in var_corr[peak_index] %}
      <tr>
        <td>
          {% if gene["gene_name"] %}
              <a href="{{ url_for("gene", gene_name=gene["gene_name"]) }}">{{ gene["gene_name"] }}</a>
          {% else %}
              {{ gene["feature_id"] }}&nbsp;
          {% endif %}
        </td>
        <td><a href="{{ url_for("gene", gene_name=gene["gene_id"]) }}">{{ gene["gene_id"] }}</a></td>
        <td>{{ gene["transcript_biotype"] }}</td>
        <td>{{ gene["n_variants"] }}</td>
        <td>{{ gene["max_corr"]|round(3) }}</td>
        <td><a class='show_genes' gene="{{ gene["gene_id"] }}" href='#show'>Variants</a></td>
      </tr>
      <tr class='{{ gene["gene_id"] }}' style="display: none;">
        <td colspan="5" style="padding:0px; width: 100%">
          <table  class="table table-hover table-striped table-condensed internal-table">
            <thead>
              <tr>
              <th>CHROM:POS</th>
              <th>REF/ALT</th>
              <th>Predicted Effect</th>
              <th>Annotation</th>
              <th>Putative Impact</th>
              <th>Correlation</th>
              </tr>
            </thead>
            <tbody style="height: inherit;">
{% for variant in gene["variant_set"] %}
      <tr class='{{ tbl_color[variant["putative_impact"]] }}'>
        <td>{{ variant["CHROM_POS"] }}</td>
        <td>{{ variant["REF"] }} / {{ variant["ALT"] }}</td>
        <td>{{ variant["hgvs_p"] }}</td>
        <td>{{ variant["annotation"] }}</td>
        <td>{{ variant["putative_impact"] }}</td>
        <td>{{ variant["correlation"]|round(3) }}</td>
      </tr>
      {% endfor %}
            </tbody>
          </table>
          <br />
        </td>
      </tr>
{% endfor %}
  </tbody>
</table>


    </div>
  {% endfor %}
  </div>


    {% include 'components/geo_map.html' %}

    {% include 'components/tajima.html' %}


  </div>

{% endif %}
{% endcache %}
{% else %}
  <div class="progress">
    <div class="progress-bar progress-bar-success progress-bar-striped active" role="progressbar"
    aria-valuenow="40" aria-valuemin="0" aria-valuemax="100" style="width:0%">
    </div>
  </div>
{% endif %}

{% endblock %}


{% block script %}
<script>

{% if trait_data.status != "complete" %}
prog = Array()
prog["queue"] = ""
prog["Processing Phenotype Data"] = "10%"
prog["Performing Mapping"] = "20%"
prog["Processing Mapping"] = "60%"
prog["Plotting Figures"] = "75%"
prog["Processing Peaks"] = "80%"
prog["Fine Mapping"] = "85%"
prog["Transferring Data"] = "95%"
prog["complete"] = "100%"
prog["error"] = "100%"

var update_status = function(status) {
  if (status == "complete") {
    $(".progress").fadeOut();
    location.reload();
  } else if (status == "error") {
    $(".progress-bar").addClass("progress-bar-danger").html("error");
  } else if (status == "queue") {
    $(".progress-bar").css("width", "100%").addClass("progress-bar-info").html("Queued")
  } else {
    $(".progress-bar").removeClass("progress-bar-info").css("width", prog[status]).html(prog[status] + " (" + status + ")");
  }
}

function check_status() {
  $.ajax({
     url:"/report_progress/",
     data: JSON.stringify({
            "report_slug": "{{ report_slug }}",
            "trait_slug": "{{ trait_slug }}" }),
     method: "POST",
     contentType: 'application/json;charset=UTF-8'
     }
     ).done(function(msg) {
      msg = JSON.parse(msg);
      update_status(msg);
    });
}

check_status();

window.setInterval(check_status, 3000);

{% else %}



// Scrolling
$('#report_nav').ready(function () {

    $('body').scrollspy({target: ".navbar", offset: 75}); 
    // Add smooth scrolling to all links inside a navbar
    $("#report_nav a").on('click', function(event){

      // Prevent default anchor click behavior
      event.preventDefault();

      // Store hash (#)
      var hash = this.hash;

      // Using jQuery's animate() method to add smooth page scroll
      // The optional number (800) specifies the number of milliseconds it takes to scroll to the specified area (the speed of the animation)
      $('html, body').animate({
        scrollTop: $(hash).offset().top - 74
      }, 800, function(){

        // Add hash (#) to URL when done scrolling (default click behavior)
        window.location.hash = hash;
      });
    });

});



{% if mapping_results|length > 0 %}
// Set Gene Counts
$(document).ready(function() {
$.each($(".confidence_intervals > tbody > tr").children(":nth-child(4)"), function(k, v) {
  pos = $(v).siblings().first().children().html();
  chrom = pos.split(":")[0]
  start = pos.split(":")[1].split("-")[0].split(",").join("");
  end = pos.split(":")[1].split("-")[1].split(",").join("");
  gene_count = $(this);
  url = "/api/interval/" + chrom + "/" + start + "/" + end;
  $.getJSON(url, function(msg) {
    $.each($(".dat_" + k), function(x) {
      try {
        $(this).html(eval($(this).attr("data")));
      } catch(err) {}
    });
});
});
});


// Gene List
function update_gene_list(chrom, start, end) {
  $("#gene_list > tbody").html("");
    $.ajax({
    url: "/api/genelist/" + chrom + "/" + start + "/" + end,
    contentType: 'application/json',
    success:function(msg) {
      $.each(msg["gene_list"], function(k,v) {

      if (v["locus"] != "") {
        v["locus"] = "<a href='http://www.wormbase.org/species/c_elegans/gene/" + v["locus"] + "'>" + v["locus"] + "</a>";
      }
      if (v["Name"] != "") {
        v["Name"] = "<a href='http://www.wormbase.org/species/c_elegans/gene/" + v["Name"] + "'>" + v["Name"] + "</a>";
      }


      items = [v["CHROM"],
               Number(v["start"]).toLocaleString('en'),
               Number(v["end"]).toLocaleString('en'),
               v["locus"],
               v["sequence_name"],
               v["Name"],
               v["biotype"]];
      $("#gene_list > tbody").append("<tr><td>" + items.join("</td><td>") + "</td></tr>");
      });

    // Enable links
    $( "#gene_list a" ).on( "click", function() {
        window.location = $(this).attr('href');
    });


    }
  });

}


$(".show_genes").on("click", function() {
  var clicked = $(this);
  console.log(clicked);
  $("." + $(this).attr("gene")).toggle(function() {})
})


// Fix Links
$( "a" ).on( "click", function() {
    window.location = $(this).attr('href');
});


// Load initial map and Tajima
$(document).ready(function() {
  update_geo_map('{{ mapping_results.0.chrom }}' , {{ mapping_results.0.pos }});
  // Load initial position
  update_tajima_plot('{{ mapping_results.0.chrom }}', {{ mapping_results.0.interval_start }}, {{ mapping_results.0.interval_end }} );
  update_gene_list('{{ mapping_results.0.chrom }}', {{ mapping_results.0.interval_start }}, {{ mapping_results.0.interval_end }});
});

// Monitor peak tabs
$('a[data-toggle="tab"]').on('shown.bs.tab', function (e) {
  peak = $(e.target).parent().attr("peak");
  chrom = peak.split(":")[0];
  pos = peak.split(":")[1];
  interval = $(".peak-tab.active").children("a").html()
  start = parseInt(interval.split(":")[1].split("-")[0]);
  end = parseInt(interval.split(":")[1].split("-")[1]);
  update_geo_map(chrom, pos);
  update_tajima_plot(chrom, start, end);
  update_gene_list(chrom, start, end);
});

{% endif %}
{% endif %}

</script>
{% endblock %}
