{% extends "_layouts/default.html" %}


{% block custom_head %}
<meta name="viewport" content="width=device-width, initial-scale=1" />
<script src="/static/report/jquery-1.11.3/jquery.min.js"></script>
<script src="/static/report/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="/static/report/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="/static/report/tocify-1.9.1/jquery.tocify.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="/static/report/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="/static/report/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="/static/report/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="/static/report/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="/static/report/htmlwidgets-0.6/htmlwidgets.js"></script>
<script src="/static/report/datatables-binding-0.1/datatables.js"></script>
<script src="/static/report/datatables-1.10.7/jquery.dataTables.min.js"></script>
<link href="/static/report/datatables-default-1.10.7/dataTables.extra.css" rel="stylesheet" />
<link href="/static/report/datatables-default-1.10.7/jquery.dataTables.min.css" rel="stylesheet" />
<link href="/static/report/datatables-responsive-1.10.7/dataTables.responsive.min.css" rel="stylesheet" />
<script src="/static/report/datatables-responsive-1.10.7/dataTables.responsive.min.js"></script>
<link href="/static/report/nouislider-7.0.10/jquery.nouislider.min.css" rel="stylesheet" />
<script src="/static/report/nouislider-7.0.10/jquery.nouislider.min.js"></script>
<link href="/static/report/selectize-0.12.0/selectize.bootstrap3.css" rel="stylesheet" />
<script src="/static/report/selectize-0.12.0/selectize.min.js"></script>

<script type="text/javascript" async
  src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-MML-AM_CHTML">
</script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/d3/3.5.16/d3.js"></script>
<script type="text/javascript" src="/static/js/d3_exploding_boxplot.js"></script>
<link href="/static/css/d3_exploding_boxplot.css" rel="stylesheet" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.0.0-alpha/js/bootstrap.min.js"></script>

{% endblock %}


{% block style %}
{% endblock %}

{% block content %}


<div class="row">
      <div class="col-md-12">

      <div class="btn-group" role="group">
        {% for trait in report_data %} 

          <a href="{{ url_for('trait_view', report_slug = report_url_slug, trait_slug = trait.trait_slug) }}" type="button" class="btnbtn-secondary btn-sm {% if trait.trait_slug == trait_slug %} btn-primary {% endif %}">{{ trait.trait_name }}</a>     
          
        {% endfor %}

        </div>
        <hr />
      </div>
</div>

<div class="row">

<div class="col-md-12">

{% if trait_slug and report_html == "" %}
<div class="progress">
  <div class="progress-bar progress-bar-success progress-bar-striped active" role="progressbar"
  aria-valuenow="40" aria-valuemin="0" aria-valuemax="100" style="width:0%">
  </div>
</div>
{% else %}

<div class="panel panel-default">
  <div class="panel-heading">
    <h3 class="panel-title">{{ trait_data["trait_name"] }}</h3>
  </div>
  <div class="panel-body">
  <div class="row">
  <div class='col-md-6'>
  <strong>Data</strong>
    <ul>
      {% if 'raw_phenotype.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/raw_phenotype.tsv">Raw Phenotype Data</a></strong></li>
      {% endif %}

      {% if 'processed_phenotype.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/processed_phenotype.tsv">Processed Phenotype Data</a></strong></li>
      {% endif %}

      {% if 'interval_variants.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/interval_variants.tsv">Interval Variants</a></strong></li>
      {% endif %}

      {% if 'raw_mapping.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/raw_mapping.tsv">Raw Mappings</a></strong></li>
      {% endif %}

      {% if 'processed_sig_mapping.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/processed_sig_mapping.tsv">Processed Significant Mappings</a></strong></li>
      {% endif %}

      {% if 'non_sig_mapping.tsv' in report_files %}
      <li><strong><a href="https://storage.googleapis.com/cendr/{{ report_slug }}/{{ trait_slug }}/tables/non_sig_mapping.tsv">Nonsignificant Mappings</a></strong></li>
      {% endif %}
    </ul>
    </div>
    <div class='col-md-6'>
    <strong>Info</strong>
    <ul>
      <li><strong>Version</strong>: {{ trait_data.version }}</li>
      <li><strong>Availibility</strong>: {% if trait_data.release == 0 %}
                                            Public
                                         {% elif trait_data.release == 1 %}
                                            Embargo
                                         {% else %}
                                            Private
                                         {% endif %}</li>
    </ul>
    </div>
    </div>
  </div>
</div>

{% endif %}

  </div>
</div>

<div id="report2" data-toggle="tab" >


<h2>Phenotype</h2>


<img src="{{ base_url }}/figures/phenotype_histogram-1.png" />
<strong>Histogram representation of your phenotype</strong>

{{ mapping_results }}
{% if mapping_results %}

  Your phenotype is significantly associated with genetic variation present in the <em>C. elegans</em> population!

{% else %}
  Unfortunately your phenotype was not significantly associated with genetic variation present in the <em>C. elegans</em> population. This could be due to noisy trait data - have you performed heritability analysis for your phenotype using our heritability strain panel? If you have and the heritability was found to be high for this trait, the trait might not have reached statistical significance because it is highly complex and more strains need to be phenotyped. If you have phenotyped the entire 152 wild-isolate collection, you can patiently wait for more isolates to be added to the collection, or generate an F2 recombinant inbred line (RIL) panel generated between strains with high and low phenotypes.
{% endif %}

<h2>Manhattan Plot</h2>

{% if mapping_results %}
<img src="{{ base_url }}/figures/Manplot-1.png" />

A genome-wide representation of the association between variation in the <em>C. elegans</em>em> population and your phenotype. The x-axis corresponds to genomic position with chromosome number indicated in the gray box above each facet. Dots represent single-nucleotide variant (SNV) present in the <em>C. elegans</em> population. The y-axis corresponds to the level of significance for the association test. Blue dots represent SNVs that are above the defined significance threshold, which is the thick gray line. Red boxes surrounding blue dots represent the QTL confidence interval, which we define as plus/minus 50 SNVs from the last signifincant SNV. The default threshold for significance is the Bonferroni-corrected value $(-log_{10}\frac{0.05}{\#SNVs})$ and is usually around 5.5.

<h3>QTL Confidence Intervals</h3>

<table class="table table-hover table-striped">
<thead>
  <tr>
    <th>Interval</th>
    <th>Peak Position</th>
    <th>-log<sub>10</sub>p</th>
  </tr>
</thead>
<tbody>
{% for row in mapping_results %}
<tr>
  <td><a href="http://www.wormbase.org/tools/genome/gbrowse/c_elegans_PRJNA13758/?name={{ row.chrom}}:{{ row.interval_start }}..{{ row.interval_end }}" target="_blank">{{ row.chrom }}:{{ "{:,.0f}".format(row.interval_start) }}-{{ "{:,.0f}".format(row.interval_end) }}</a></td>
  <td><a href="http://www.wormbase.org/tools/genome/gbrowse/c_elegans_PRJNA13758/?name={{ row.chrom}}:{{ row.interval_start }}..{{ row.interval_end }}" target="_blank">{{row.chrom}}:{{ "{:,.0f}".format(row.pos) }}</a></td>
  <td>{{ row.log10p|round(3) }}</td>
</tr>
{% endfor %}
</tbody>
</table>
<h3>QQ-Plot</h3>

<img src="{{ base_url }}/figures/QQplot-1.png" />
<p>Quantile-quantile plot generated from the p-values from the association mapping. This plot shows the expected distribution of association test statistics on the x-axis compared to the observed values on the y-axis. Each dot corresponds to an individual SNV. SNVs that deviate from the x=y line indicate that they are associated with your phenotype. Earlier separation from the identity line is indication of population stratification not being eliminated by the mapping algorithm.</p>

<h3>Phenotype-by-Genotype</h3>


<p>Phenotypic distributions represeted as box plots are split by the genotype at the most significant SNV for a given QTL. The phenotype is on the y-axis and the genotype is on the x-axis. The chromosome and position for the plotted SNV are indicated in each the gray rectangles above the panels. REF refers to the N2 genotype and ALT refers to the variant phenotype. All of the SNVs in our data set are biallelic so these are the only two classes for any given SNV.</p>

<div id="chartContainer"></div>


{% else %}
{# nonsignificant mapping #}

{% endif %}

<div>



<div id="report">
            {{ report_html|safe }}
</div>
<style>

#TOC {
  width: 100%;
}

div.tocify {
  width: 100%;
  max-width: 240px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

#report .main-container {
  padding: 0px;
}

</style>


{% endblock %}


{% block script %}

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}
});
</script>

<script>


//generate random data
var box1 = d3.range(100).map(d3.random.normal(Math.random()*100,5))
            .map(function(d,i){return {v:d,g:'box1',t:'pt 1'+i}})

var box2 = d3.range(100).map(d3.random.normal(Math.random()*100,30))
            .map(function(d,i){return {v:d,g:'box2',t:'pt 2'+i}})


var data = box1.concat(box2)

// chart(data,aes)
// aesthetic :
// y : point's value on y axis
// group : how to group data on x axis
// color : color of the point / boxplot
// label : displayed text in toolbox
var chart = exploding_boxplot(data,
            {y:'v',group:'g',color:'g',label:'t'})

//call chart on a div
chart('#chartContainer')

$(document).ready(function () {
$('svg circle').tooltip({
      'placement': 'right'
  });
});
  
</script>

<script>

{% if trait_slug and report_html == "" %}
prog = Array()
prog["queue"] = ""
prog["Processing Phenotype Data"] = "10%"
prog["Performing Mapping"] = "20%"
prog["Processing Mapping"] = "60%"
prog["Plotting Figures"] = "75%"
prog["Processing Peaks"] = "80%"
prog["Transferring Data"] = "95%"
prog["complete"] = "100%"
prog["error"] = "100%"


var update_status = function(status) {
  if (status == "complete") {
    $(".progress").fadeOut();
    location.reload();
  } else if (status == "error") {
    $(".progress-bar").addClass("progress-bar-danger").html("error");
  } else if (status == "queue") {
    $(".progress-bar").css("width", "100%").addClass("progress-bar-info").html("Queued")
  } else {
    $(".progress-bar").removeClass("progress-bar-info").css("width", prog[status]).html(prog[status] + " (" + status + ")");
  }
}

function check_status() {
  $.ajax({
     url:"/report_progress/",
     data: JSON.stringify({
            "report_slug": "{{ report_slug }}",
            "trait_slug": "{{ trait_slug }}" }),
     method: "POST",
     contentType: 'application/json;charset=UTF-8'
     }
     ).done(function(msg) {
      msg = JSON.parse(msg);
      update_status(msg);
    });
}

check_status();

window.setInterval(check_status, 3000);
{% endif %}

</script>
{% endblock %}
