{% extends "_layouts/default.html" %}

{% block content %}
<div class="row">
      <div class="col-md-6">
      
<img src="{{ gs_static('img/crawl.gif') }}" class='img-responsive' />
<div style="text-align:center">
<small>Movie of <em>C. elegans</em> taken by the <a href="https://goldsteinlab.weebly.com/">Goldstein Lab</a></small>
</div>

      </div>
      <div class="col-md-6">

{% filter markdown %}

### What is _C. elegans_?

_Caenorhabditis elegans_ is a non-parasitic nematode roundworm that lives in rotting material and eats bacteria and fungi. Because this species grows easily and quickly in the laboratory, it is a powerful model to learn about human development, complex behaviors, and evolutionary processes. For more information about _C. elegans_, [see this wikipedia page](https://en.wikipedia.org/wiki/Caenorhabditis_elegans) or learn about its history at [wormclassroom.org](http://wormclassroom.org/short-history-c-elegans-research).

{% endfilter %}
      </div>
  </div>
  <div class="row">

  <div class="col-md-6">



  <h3>Global Distribution of wild isolates</h3>
<p>Most research groups that study <em>C. elegans</em> focus on the laboratory-adapted strain (called N2) isolated in Bristol, England in the 1950s. We have learned a great deal about basic biological processes from studies of this one strain.</p>
<p>
However, this species is found worldwide, and wild strains are as different from one another as humans are different from one another. These strains are isolated from a variety of environments in nature when researchers collect rotting materials, including fruits, flowers, nuts, berries, stems, leaves, and compost. We can use the natural diversity of these strains to learn about how populations of individuals are genetically different from another and how those differences might impact disease. </p>

  </div>


  <div class="col-md-6">
<div id="map" class="col-md-8" style="height: 500px; width:100%">
  <script>
    mapboxgl.accessToken = 'pk.eyJ1Ijoic3dhY2hzcHJlc3MiLCJhIjoiY2trZDJyZWtmMHM3OTJvcWlpNmE3cTQ1ZSJ9.UwY-WYd78c7SnO1S_37vtg';
    var map = new mapboxgl.Map({
        container: 'map',
        style: 'mapbox://styles/mapbox/outdoors-v11', // stylesheet location
        center: [ '0', '0'], // starting position [lng, lat]
        zoom: -2 // starting zoom
    });
  </script>
</div>
  </div>
  </div>
    <div class="row">
    <div class="col-md-6">
      <div class="row">
      <div class="col-md-6">
          <img src="/static/img/logo.png" style="max-width: 180px;" class="img-responsive" />
      </div>
      <div class="col-md-6">
          <img src="/static/img/logos/NU.png" style="top:15px; position: relative;" class="img-responsive"/>
      </div>
      </div>
    </div>


<div class="col-md-6">

<h3>CeNDR Goals</h3>
<p>To facilitate the study of natural diversity by <em>C. elegans</em> research groups, we created the <em>C. elegans</em> Natural Diversity Resource (CeNDR). We have three major goals:</p>

<ol>
<li>To accept, organize, and distribute wild strains to research groups that want to investigate their favorite trait(s) across natural <em>C. elegans</em> strains. See <strong><a href="{{ url_for('strains.strains_catalog') }}">Strains</a></strong>.</li>

<li>To sequence the whole genomes of wild <em>C. elegans</em> strains, provide the aligned sequence data, and facilitate discovery of genetic variation across the entire species. See <strong><a href="{{ url_for('data.data') }}">Data</a></strong>.</li>

<li>To perform genome-wide association mappings to correlate genotype with phenotype and identify genetic variation underlying quantitative traits. </li>
</ol>

<p>Please <a href="{{ url_for('primary.contact') }}">let us know</a> what we can do to facilitate your discoveries! We are interested in adding new resources and tools.</p>

</div>
  </div>
{% endblock %}



{% block script %}

<script>
  if (document.getElementById('map')) {
    ACCESS_TOKEN = mapboxgl.accessToken;
    var markerHeight = 50, markerRadius = 10, linearOffset = 25;
    var popupOffset = {
        'top': [0, 0],
        'top-left': [0,0],
        'top-right': [0,0],
        'bottom': [0, -markerHeight],
        'bottom-left': [linearOffset, (markerHeight - markerRadius + linearOffset) * -1],
        'bottom-right': [-linearOffset, (markerHeight - markerRadius + linearOffset) * -1],
        'left': [markerRadius, (markerHeight - markerRadius) * -1],
        'right': [-markerRadius, (markerHeight - markerRadius) * -1]
    };

  data = {{ strain_listing|tojson|safe }}

    data.forEach(function(d) {
        if (d.latitude) {
            const innerHTML = `
                <div>
                    <ul style="list-style-type:none; padding-left:0px">
                        ${d.isotype ? '<li><b><a href="../strains/isotype/' + d.isotype + '">Isotype: ' + d.isotype + '</a></b></li>': ''}
                        ${d.strain ? '<li>Strain: ' + d.strain + '</li>': ''}
                        ${d.isolation ? '<li>Isolation: ' + d.isolation + '</li>': ''}
                        ${d.elevation ? '<li>Elevation: ' + d.elevation + '</li>': ''}
                        ${d.isolation ? '<li>Location: ' + d.location + '</li>': ''}
                    </ul>
                </div>
            `;
            p = new mapboxgl.Popup()
                .setHTML(innerHTML)
                .setOffset(popupOffset);
            m = new mapboxgl.Marker()
                .setLngLat([d.longitude, d.latitude])
                .setPopup(p)
                .addTo(map);
        }
    })
}
</script>

{% endblock %}