{% extends "_layouts/default.html" %}

{% block custom_head %}
	<link rel="stylesheet" href="//cdn.datatables.net/1.10.21/css/jquery.dataTables.min.css" />
	<script type="text/javascript" src="//cdn.datatables.net/1.10.21/js/jquery.dataTables.min.js"></script>
	<script type="text/javascript" src="//igv.org/web/release/2.6.2/dist/igv.min.js"></script>
{% endblock %}


{% block content %}
{% from "macros.html" import render_field %}

<div class="container-fluid">
	<div class = "row">
		<p>
			This web-based tool is designed to compare any two wild <i>C. elegans</i> strains for insertion-deletion (indel) variants that can be genotyped using PCR. These molecular markers can be used to follow each respective genetic background in crosses. Enter a specific genomic region. The following page will output indel variants that can be genotyped by PCR. For each of these indels, you can search for primers to genotype them in the two wild strains using differential PCR
			product sizes. Primers are designed to avoid natural variants in either strain to ensure that the PCR works in both genetic backgrounds.
		</p>
	</div>
	<div class='row'>
		<form id='form-submit' method="POST">
			{{ form.csrf_token }}
			<div class='col-md-2'>{{ render_field(form.strain_1) }}</div>
			<div class='col-md-2'>{{ render_field(form.strain_2) }}</div>
			<div class='col-md-2'>{{ render_field(form.chromosome) }}</div>
			<div class='col-md-2'>{{ render_field(form.start, placeholder="1,425,000", value=1) }}</div>
			<div class='col-md-2'>{{ render_field(form.stop, placeholder="1,465,000", value=1000000) }}</div>
			<div class='col-md-2'>
				<label>&nbsp;</label> <br />
				<button id='find_indels' type='submit' value="submit" class="btn btn-primary btn-block"><span class="glyphicon glyphicon-search" aria-hidden="true"></span> Find Indels</button>
			</div>
		</form>
	</div>
	<div class="row">
		<div id="browser"></div>
		<table id="results" class='table table-striped table-hover table-compact'></table>
	</div>
</div>

{% endblock %}

{% block script %}

<script>

/*
	IGV Browser
*/
var browserDiv = $("#browser"),
                options = {
                    showNavigation: true,
                    showKaryo: false,
                    reference: {
                      id: "WS245",
                      fastaURL: "//storage.googleapis.com/elegansvariation.org/browser_tracks/c_elegans.PRJNA13758.WS245.genomic.fa",
                    },
                };

var loaded = false
var table
$(document).ready(function(){
	
	$("#browser").on("click", ".ortholink", function() {
		igv.getBrowser().search($(this).attr("link"))
		// Update tables and URL
		refresh_variants();
	  });

	set_position = function() {
		// Sets browser position based on input
		browser = igv.getBrowser()
		chromosome = $("#chromosome").val()
		start = $("#start").val()
		end = $("#stop").val()
		console.log(`${chromosome}:${start}-${end}`)
		console.log(browser)
		browser.search(`${chromosome}:${start}-${end}`)
	}

	submitForm = function() {
		$("form").submit()
	}

    $('form').submit(function (e) {
		e.preventDefault();
		//DA = { id1: $("#strain1 option:selected").text(), id2: $("#strain2 option:selected").text(), chrom: $("#chrom option:selected").text(), start:$("#di").val(), stop:$("#ti").val()}
		var url = "{{ url_for('tools.pairwise_indel_finder_query') }}"
		$.ajax({type: "POST",
				url: url,
				data: $('form').serialize(),
				success: function(data) {
					results = data["results"]
					console.log(results)
					// Handle errors
					$(".form-error").remove();
					if (Object.keys(data).indexOf("errors") > -1) {
						for(error in data["errors"]) {
							$("[for='" + error + "']").parent().append(`<div class='form-error'>${data["errors"][error]}</div>`)
						}
					} else {
						if (loaded == false) {
							// Create genome browser
							browser = igv.createBrowser(browserDiv, options)
							             .then(function(browser) {
											 set_position()

											 browser.on('locuschange', function(reference_frame, label) {
												 $("#chromosome").val(reference_frame["chr"]);
												 $("#start").val(reference_frame["start"]);
												 $("#stop").val(reference_frame["end"]);
												 setTimeout(submitForm, 2000)
											 })
										 })
							tracks = {}
							// Restructure results by strain
							for(idx in results) {
								row = results[idx]
								if (row["STRAIN"] in tracks == false) {
									tracks[row["STRAIN"]] = []
								}
								tracks[row["STRAIN"]].push(row)
							}


							table = $("#results").DataTable({
								data: results,
								paging:   false,
								ordering: true,
								searching: false,
								order: [[1, "asc"]],
								columnDefs: [
									{ targets: [0], orderData: [1,2] }
								],
								columns: [
								{ data: "site", title: "Site" },
								{ data: "START", title: "Start", visible: false },
								{ data: "END", title: "End", visible: false },
								{ data: "STRAIN", title: "Strain" },
								{ data: "overlap", title: "overlap"}
								]
							});
							loaded = true;
						} else {
							table.clear()
							table.rows.add(results);
							table.draw();
						}
					}
				}
			});
	});

	$.ajaxSetup({
		beforeSend: function(xhr, settings) {
			if (!/^(GET|HEAD|OPTIONS|TRACE)$/i.test(settings.type) && !this.crossDomain) {
				xhr.setRequestHeader("X-CSRFToken", "{{ form.csrf_token._value() }}")
			}
		}
	})
});
</script>

{% endblock %}


