{% extends "_layouts/default.html" %}


{% block custom_head %}

<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta http-equiv="X-UA-Compatible" content="IE=edge" />

{% endblock %}

{% block style %}
<style>
.report {
  width: 100%;
} 

.height-report {
  height: 60%;
  width: 60%;
  display: block;
  margin-left: auto;
  margin-right: auto
}

#report_nav.affix {
    position: fixed;
    top: 0px;
    left: 0px;
    width: 100%;
}

/* Keep Header Scroll */
#variant_list table {
    width: 100%;
}

#variant_list thead, #variant_list tbody, #variant_list tr, #variant_list td, #variant_list th { display: block; }

#variant_list tr:after {
    content: ' ';
    display: block;
    visibility: hidden;
    clear: both;
}

#gene_list thead th, #variant_list thead th {
    height: 30px;
    /*text-align: left;*/
}

#gene_list tbody, #variant_list tbody {
    height: 700px;
    overflow-y: auto;
}

#gene_list tbody td, #gene_list thead th, #variant_list tbody td, #variant_list thead th {
    width: 16.66%;
    float: left;
}

.meta-table td:nth-child(1) {
    font-weight: bold;
}

.meta-table td:nth-child(2) {
    text-align: right;
}

.corr-row td:nth-child(1) {
    background-color: inherit !important;
}

</style>
{% endblock %}

{% block content %}

<div class="row">
    <div class='col-md-12'>
        <h4>Traits</h4>
        {# Traits #}
            {% for trait in report.trait_list %} 
                <a href="{{ url_for('mapping.report', report_slug = report.report_slug, trait_name = trait) }}" type="button" class="btn btn-sm {% if trait_name == trait %} btn-primary {% else %} btn-link {% endif %}"><strong>{{ trait }}</strong></a>
            {% endfor %}
        <br /><br />
    </div>{# /col-md-12 #}
</div>{# /row #}


<div class="panel panel-default">
    <div class="panel-body">
        <div class="row">
            <div class='col-md-12'>
            <h5>Report summary</h5>
            </div>
        </div>
        <div class='row'>
            <div class='col-md-6'>
                {% from "macros.html" import render_table %}
                {% set rows = [['Report Name', report.report_name],
                               ['Trait Name', trait.trait_name],
                               ['Data release', trait.version_link()],
                               ['CeNDR version', trait.CENDR_VERSION],
                               ['Report version', trait.REPORT_VERSION],
                               ['cegwas version','']] %}
                {{ render_table(header=None, rows=rows, table_class='table-striped table-bordered meta-table') }}

                {% set rows = [['Strains for trait "{}"'.format(trait.trait_name), strain_count],
                               ['Traits', report.trait_list|length]] %}
                {{ render_table(header=None, rows=rows, table_class='table-striped table-bordered') }}
            </div>{# /col-md-6 #}
            <div class='col-md-6'>

                <table class='table table-striped table-bordered meta-table'>
                    <tbody>
                        <tr>
                            <td>Status</td>
                            <td>{{ trait.run_status }}</td>
                        </tr>
                        <tr>
                            <td>Submitted</td>
                            <td>{{ trait.created_on.strftime("%Y-%m-%d&nbsp;&nbsp; %H:%m:%S")|safe }} UTC</td>
                        </tr>
                        <tr>
                            <td>Started</td>
                            <td>{% if trait.started_on %}
                                {{ trait.started_on.strftime("%Y-%m-%d&nbsp;&nbsp; %H:%m:%S")|safe }} UTC
                            {% endif %}</td>
                        </tr>
                        <tr>
                            <td>Completed </td>
                            <td>{% if trait.completed_on %}
                                {{ trait.completed_on.strftime("%Y-%m-%d&nbsp;&nbsp; %H:%m:%S")|safe }} UTC
                            {% endif %}</td>
                        </tr>
                        <tr>
                            {% set duration = trait.duration() %}
                            <td>Duration</td>
                            <td>{% if duration %}{{ trait.duration() }}{% endif %}</td>
                        </tr>
                        <tr>
                            <td>Log</td>
                            <td><a href='javascript:$(".log").slideToggle();'>View Log</a></td>
                        </tr>
                    </body>
                </table>


                <table class='table table-striped table-bordered meta-table'>
                    <tbody>
                        <tr>
                            <td>Report Mode</td>
                            <td>{% if report.is_public %}public{% else %}private{% endif %}</td>
                        </tr>
                        <tr>
                            <td>Share</td>
                            <td>{% if report.is_public %}
                                    {{ request.url_root.strip('/') }}{{ url_for('mapping.report', report_slug=report.report_slug, trait_name=trait_name) }}
                                {% else %}
                                    {{ request.url_root.strip('/') }}{{ url_for('mapping.report', report_slug=report.secret_hash, trait_name=trait_name) }}
                                {% endif %}
                            </td>
                        </tr>
                    </tbody>
                </table>

            </div>
        </div>{# /row #}

        <div class="row">
            <div class='col-md-12'>
            <h5>Data</h5>
            </div>
        </div>
        <div class='row'>
            <div class='col-md-6'>

            </div>
        </div>


    </div>{# /panel-body #}
</div>{# /panel #}

<div id="report" data-toggle="tab">
    <nav id="report_nav" data-offset-top="385" data-spy="affix" class="navbar navbar-default" style="z-index:10000">
      <div class="container-fluid">
       <!-- Collect the nav links, forms, and other content for toggling -->
        <div class="collapse navbar-collapse">
          <ul class="nav navbar-nav">
            <li><a href="#phenotype">Phenotype</a></li>
            <li><a href="#manhattan">Manhattan</a></li>
            {% if mapping_results|length > 0 %}
            <li><a href="#intervals">Intervals</a></li>
            <li><a href="#pxg">PxG</a></li>
            {% if mapping_results|length > 1 %}
            <li><a href="#ld">LD</a></li>
            {% endif %}
            <li><a href="#peak_summary">Peak Summary</a></li>
            {% endif %}
          </ul>
        </div><!-- /.navbar-collapse -->
      </div><!-- /.container-fluid -->
    </nav>

    <div class='log'>
        <h4>Log</h4>
        <div class='well'>
            {% for line in trait.get_formatted_task_log() %}
                {{ line }}<br />
            {% endfor %}
        </div>
    </div>


    <h2 id="phenotype" name="phenotype">Phenotype</h2> 
    {{ phenotype_plot|safe }}

    {% if trait.is_significant %}

      Your phenotype is significantly associated with genetic variation present in the <em>C. elegans</em> population!

    {% else %}

      <p>Unfortunately your phenotype was not significantly associated with genetic variation present in the <em>C. elegans</em> population. This could be due to noisy trait data - have you performed heritability analysis for your phenotype using our heritability strain panel? If you have and the heritability was found to be high for this trait, the trait might not have reached statistical significance because it is highly complex and more strains need to be phenotyped.</p>

      <p>If you have phenotyped the entire 152 wild-isolate collection, you can patiently wait for more isolates to be added to the collection, or generate an F2 recombinant inbred line (RIL) panel generated between strains with high and low phenotypes.</p>

    {% endif %}

    <h2 id="manhattan" name="Manhattan">Manhattan</h2> 

    <img class='img img-responsive'src='{{ trait.file_url("Manhattan.png") }}' alt='Manhattan Plot' />

    {% if trait.is_significant %}

    <p>A genome-wide representation of the association between variation in the <em>C. elegans</em> population and your phenotype. The x-axis corresponds to genomic position with chromosome number indicated as a roman numeral above each box. Dots represent single-nucleotide variants (SNV) present in the <em>C. elegans</em> population. The y-axis corresponds to the level of significance for the association test. Blue dots represent SNVs that are above the defined significance threshold, which is the thick gray line. Red boxes surrounding blue dots represent the QTL genomic region of interest, which we define as plus/minus 50 SNVs from the last signifincant SNV. The default threshold for significance is the Bonferroni-corrected value <img src='/static/img/report/bf.png' style="height: 30px;" /> and is usually around 5.5.</p>

    {% else %}

    <p>A genome-wide representation of the association between variation in the <em>C. elegans</em> population and your phenotype. The x-axis corresponds to genomic position with chromosome number indicated as a roman numeral above each box. Dots represent single-nucleotide variants (SNV) present in the <em>C. elegans</em> population. The y-axis corresponds to the level of significance for the association test. The red line corresponds to the Bonferroni-corrected significance threshold, which is defined as <img src='/static/img/report/bf.png' style="height: 30px;" />.</p>

    {% endif %}

    <h2 id="pxg">Phenotype-by-Genotype</h2>

    {{ pxg_plot|safe }}

    <p>Phenotypic distributions represented as box plots are split by the genotype at the most significant SNV for a given QTL. The phenotype is on the y-axis, and the genotype is on the x-axis. The chromosome and position for the plotted SNV are indicated as roman numerals above each panel. REF refers to the reference N2 genotype, and ALT refers to the alternative variant phenotype. </p>


    {% if mapping_results|length > 1 %}
    <h2 id="ld">Linkage Disequilibrium</h2>
    <div class='row'>
        <div class='col-md-8 col-md-offset-2 text-center'>
            <br />
            <img class='img img-responsive' src='{{ trait.file_url("LD.png") }}' alt='LD Plot' />
            <br />
            <img style='width: 50%;' src='https://elegansvariation.org/static/img/report/tajimad.png' alt="LD Calc" />
            <br />
        </div>
    </div>
    <p>The Linkage Disequilibrium (LD) measure plotted is the correlation between peak markers.</p>
    <p>D is the raw difference in frequency between the observed number of AB pairs and the expected number and A, B, a, and b refer to alleles at the two loci. An LD value of 0.8 or higher suggests that the two peak markers are not segregating randomly.</p>


    {% endif %}

    <h2 id="peak_summary">Peak Summary</h2>

    <p><strong>Switch between peaks below.</strong></p>

    <!-- Nav tabs -->
    <ul class="nav nav-tabs" role="tablist">
    {% for peak in mapping_results.to_dict('records') %}
        <li role="presentation" peak="{{ peak.peak_pos }}" class='peak-tab {% if loop.first %}active{% endif %}'><a href="#peak_{{ loop.index }}" aria-controls="peak_{{ loop.index }}" role="tab" data-toggle="tab">{{ peak.interval }}</a></li>
    {% endfor %}
    </ul>

    <!-- Tab panes -->
    <div class="tab-content">
        {{ interval_summary }}
    {% for peak in mapping_results.to_dict('records') %}
    {% set peak_index = loop.index0 %}
    <div role="tabpanel" class="tab-pane fade {% if loop.first %}in active{% endif %}" id="peak_{{ loop.index }}">
    <br />
    <br />
    <ul>
      <li><strong>Peak Interval</strong> &nbsp;{{ peak.interval }}</li>
      <li><strong>Peak Position</strong> &nbsp;{{ peak.peak_pos }}</li>
      <li><strong>Variance Explained</strong> &nbsp;{{ peak.variance_explained|round(3) * 100 }}%</li>
    </ul>
    <div class="row">
        <div class="col-md-12">
          <table class="peak_summary table table-hover table-striped" style="table-layout: fixed">
                <thead>
                <tr>
                    <th>Biotype</th>
                    <th># Genes</th>
                    <th># Genes w/ variants</th>
                    <th>>=1 Moderate Impact</th>
                    <th>>=1 High Impact</th>
                    <th># Variants</th>
                </tr>
                </thead>
                <tbody class='peak_{{ loop.index0 }} text-right'>

                    {{ gene }}
                    <tr>
                        <td class='text-left'><strong>Total</strong></td>
                        <td>{{ genes_w_variants|default("-") }}</td>
                        <td class='dat_{{ peak_index }}'>-</td>
                        <td class='dat_{{ peak_index }}'>-</td>
                        <td class='dat_{{ peak_index }}'>-</td>
                        <td class='dat_{{ peak_index }}'>-</td>
                    </tr>
                  {% for biotype in ["protein_coding", "pseudogene", "ncRNA", "miRNA", "piRNA", "tRNA", "lincRNA", "rRNA", "scRNA", "snoRNA", "snRNA", "asRNA", "Coding", "Noncoding"] %}
                  <tr>
                    <td class='text-left'>
                    {% if biotype in BIOTYPES.keys() %}
                      <abbr data-toggle='tooltip' data-placement='top' data-original-title='{{ BIOTYPES[biotype] }}'>{{ biotype }}</abbr>
                    {% else %}
                    {{ biotype }}
                    {% endif %}
                    </td>
                    <td data='msg.gene_biotype.biotype_summary.{{ i }}' class='dat_{{ peak_index }}'>-</td>
                    <td data='msg.gene_w_variants_biotype.{{ i }}' class='dat_{{ peak_index }}'>-</td>
                    <td data='msg.gene_by_impact.MODERATE.{{ i }}' class='dat_{{ peak_index }}'>-</td>
                    <td data='msg.gene_by_impact.HIGH.{{ i }}' class='dat_{{ peak_index }}'>-</td>
                    <td data='msg.variant_in_biotype.{{ i }}' class='dat_{{ peak_index }}'>-</td>
                  </tr>
                  {% endfor %}
                  <tr>
                    <td colspan = "6">
                    <p style="text-align">Numbers listed under Moderate and High Impact correspond to the number of genes within the interval that have at least 1 variant classified as having a moderate or high impact.</p>
                    <p style="text-align: justify">Moderate and High impact variants are terms used to describe the predicted severity of a variant as assigned by SnpEff. More information regarding the types of variants classified as Moderate or High impact is <a href="{{ url_for('primary.help_item', filename = 'Variant-Prediction') }}">available in the help section</a>.</p>
                    </td>
                  </tr>
            </tbody>
        </table>
        </div>
    </div>
      
    <h3>Variants</h3>

    <p>The table below lists variants, their predicted effects, and their correlation with the phenotype. Only protein coding genes have their variants correlated with phenotype.</p>

    <strong><a href='{{ trait.file_url("interval_variants.tsv.gz") }}'>Download Table</a></strong>
    <table class="table table-hover table-striped" id="variant_list" style="border: 1px solid #b3b3b3;">
        <thead>
            <tr>
                <th>Gene</th>
                <th><strong>CHROM:POS (REF/ALT)</strong></th>
                <th><strong>Impact</strong></th>
                <th><strong>Effect</strong></th>
                <th><strong>AA Change</strong></th>
                <th><abbr data-toggle='tooltip' data-placement='top' data-original-title="The maximum p-value for variants in each gene from Spearman's Rank-Sum correlation test with trait data">-log10( p-value )</abbr></th>
            </tr>
        </thead>
        <tbody>
            {% for row in variant_correlation[variant_correlation.peak == peak.interval].to_dict('records') %}
                {% if loop.changed(row['gene_id']) %}
                <tr>
                    <td colspan="6" style='width: 100%; border-top: 2px ridge #999999;'>
                        <div class='pull-right'>
                            Max p-value = {{ row["max_gene_corr_p"]|round(3) }} <br />
                            {{ row['n_variants'] }} variant(s)
                        </div>
                        {% if row["gene_name"] %}
                            <strong>
                                <a href="{{ url_for("gene.gene", gene_name=row['gene_id']) }}">
                                    {{ row["gene_name"] }}
                                </a>
                            </strong>
                            <br /> {{ row["gene_id"] }}</a>
                        {% else %}
                            <strong><a href="{{ url_for("gene.gene", gene_name=row['gene_id']) }}">{{ row["gene_id"] }}</a></strong>
                        {% endif %} -- <em>{{ row['transcript_biotype'] }}</em>
                        <br />
                    </td>
                </tr>
                {% endif %}
                <tr class='corr-row {{ TABLE_COLORS[row["impact"]] }} tr_row'>
                    <td></td>
                    <td>{{ row['CHROM'] }}:{{ row['POS'] }} ({{ row["REF"] }} / {{ row["ALT"] }})</td>
                    <td>{{ row["impact"] }}</td>
                    <td>{{ row["effect"] }}</td>
                    <td>{{ row["aa_change"] }}</td>
                    <td>{{ row["corrected_spearman_cor_p"]|round(3) }}</td>
                </tr>
            {% endfor %}
        </tbody>
    </table>

  {% endfor %}
  </div>


    {% include 'reports/geo_map.html' %}

    {% include 'reports/tajima.html' %}


  </div>


</div>
{% endblock %}

{% block script %}
<script>
var offsetHeight = 71;
// Scrolling
$('body').scrollspy({offset: 100}); 
// Add smooth scrolling to all links inside a navbar
  $("#report_nav li a").on('click', function(event){

  //event.preventDefault();
  link = $(this).attr('href').replace("#","")
  var scrollPos = $(`[id='${link}']`).offset().top - (offsetHeight - 1);
  $('html, body').animate({
    scrollTop: scrollPos
  }, 800, function(){

    return false;
  });
});

// Tab switching

// Load initial map and Tajima
var isotypes = {{ isotypes|tojson }}
var peak = "{{ first_peak.peak_pos }}";
var chrom = peak.split(":")[0]
var pos = parseInt(peak.split(":")[1])
$(document).ready(function() {
    update_geo_map();

    // Load initial position

});

// Monitor peak tabs
$('a[data-toggle="tab"]').on('shown.bs.tab', function (e) {
  peak = $(e.target).parent().attr("peak");
  chrom = peak.split(":")[0];
  pos = peak.split(":")[1];
  interval = $(".peak-tab.active").children("a").html()
  start = parseInt(interval.split(":")[1].split("-")[0]);
  end = parseInt(interval.split(":")[1].split("-")[1]);
  update_geo_map(chrom, pos);
  update_tajima_plot(chrom, start, end);
  update_gene_list(chrom, start, end);
});


</script>
{% endblock %}
